---
title: "hvo_mating"
author: "Artemis S. Louyakis"
date: "6/8/2020"
output: html_document
---

*Haloferax volcanii* was grown over 24 hours in five increments - 0, 2, 4, 8, and 24 - with a non-mating, shaking control, p. 

Programs:
https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_Manhattan_Plots_in_R

```{r load packages, echo=FALSE}
# Import packages 
library(NOISeq)
library(manhattanly)
library(ggplot2)
library(DESeq2)
library(edgeR)
library(reshape2)
library(Rmisc)
library(tidyr)
library(ggpubr)
library(tidyverse)
library(ggsignif)
library(dplyr)
library(plotly)
library(factoextra)
library(pheatmap)
library(forcats)
library(cowplot)
library(phyloseq)
library(gggenes)
library(DEGreport)
library(pheatmap)
library(tibble)
library(stringr)
```

###\n Import count matrix
```{r import counts, echo=FALSE}
# Import counts data
count_data_raw <- read.table("data/mating.mat", header=T, com='', check.names=F)
count_data <- count_data_raw[rowSums(count_data_raw) > 10,]
count_data <- round(count_data, 0)
```
###\n Prepare sample metadata
```{r factors, echo=FALSE}
# Make factors for treatments 
myfactors = data.frame("heading" = substr(colnames(count_data), start=1, stop=9),
                       "rep" = substr(colnames(count_data), start=1, stop=1),
                       "time" = substr(colnames(count_data), start=2, stop=3),
                       "samp" = substr(colnames(count_data), start=1, stop=3),
                       "time_de" = c("shaking", "t00", "t02", "t04", "t08", "t24", "shaking", "t00", "t02", "t04", "t08", "t24"))
rownames(myfactors) <- colnames(count_data)
save(myfactors, file = "noiseq/myfactors_mate.RData")

# Make list of samples - a=rep one; b=rep two; ##=hours of growth; p=shaking (no mating)
samps <- c("ap0","a00","a02","a04","a08","a24","bp0","b00","b02","b04","b08","b24")
time <- c("p0","00","02","04","08","24")
rep1 <- c("ap0_quant","a00_quant","a02_quant","a04_quant","a08_quant","a24_quant")
rep2 <- c("bp0_quant","b00_quant","b02_quant","b04_quant","b08_quant","b24_quant")
```
###\n Construct annotation map
```{r annotations, echo=FALSE}
# Import annotation map
annomap <- read.delim("annotations/annomap.txt", header=F)
  annomap <- dplyr::rename(annomap, ncbi_id = V1, hvo = V2, protein = V3) #dplyr

locus_tag <- read.delim("annotations/locus_tag_map.txt", header=F)
  locus_tag <- dplyr::rename(locus_tag, hvo = V1, old_locus_tag = V2) #dplyr
keggmap <- read.delim("annotations/hvo_khier_old_locus.txt", header=F)
  keggmap <- dplyr::rename(keggmap, kegg1 = V1, kegg2 = V2, kegg3 = V3, 
                           old_locus_tag = V4, gene_id = V5, gene_description = V6,
                           kegg_id = V7, kegg_gene = V8, kegg_description = V9, enzyme = V10) #dplyr
  keggmap <- merge(keggmap, locus_tag, by="old_locus_tag", all.y=TRUE)
# kegg1 <- split(keggmap, keggmap$kegg1)
# kegg1 <- split(kegg1, keggmap$kegg2)


go_plus <- read.delim("annotations/hvo_goterms.txt", header=TRUE)

cds_ids <- read.delim("annotations/cds_ids.txt", header=FALSE)
  cds_ids <- dplyr::rename(cds_ids, ncbi_id = V1, locus_tag = V2, dbxref = V3)
  cds_ids$locus_tag <- NULL
gff_clean <- read.delim("annotations/gff_clean.txt", header=FALSE)
  gff_clean <- dplyr::rename(gff_clean, genomic_accession = V1, makenull1 = V2, feature = V3, start = V4, stop = V5, 
                      makenull2 = V6, strand = V7, makenull3 = V8, gene = V9, dbxref = V10, gbkey = V11, 
                      gene_biotype = V12, locus_tag = V13, old_locus_tag = V14)
  gff_clean$old_locus_tag <- sub("^$", NA, gff_clean$old_locus_tag)
  gff_clean$makenull1 <- NULL
  gff_clean$makenull2 <- NULL
  gff_clean$makenull3 <- NULL
hvo_map <- merge(cds_ids, gff_clean, by = "dbxref") ## tRNA and rRNA don't have ids, so not part of de analysis
hvo_map$length <- hvo_map$stop - hvo_map$start
hvo_gene_lengths <- data.frame("ncbi_id" = hvo_map$ncbi_id,
                               "length" = hvo_map$length)
hvo_map <- merge(hvo_map, annomap, by = "ncbi_id")
#hvo_map <- merge(hvo_map, sub.keggmap, by.x = "locus_tag", by.y = "hvo", all.x=TRUE)
hvo_map <- merge(hvo_map, keggmap, by.x = "locus_tag", by.y = "hvo", all.x=TRUE)
hvo_map$keggidprot <- paste0(hvo_map$kegg_id, ": ", hvo_map$kegg_prot)
hvo_map$Group.1 <- NULL
hvo_map$old <- NULL
hvo_map <- merge(hvo_map, go_plus, by.x = "dbxref", by.y = "ID", all.x = TRUE)
hvo_map$hvo <- NULL
#hvo_map <- merge(hvo_map, arcog, by = "locus_tag", all.x = TRUE)
hvo_map$old_locus_tag[is.na(hvo_map$old_locus_tag)] <- as.character(hvo_map$locus_tag[is.na(hvo_map$old_locus_tag)])

hvo_map <- within(hvo_map, {
    chr = ifelse(genomic_accession == "NC_013967.1", "Chromosome", 
                 ifelse(genomic_accession == "NC_013968.1", "pHV1", 
                        ifelse(genomic_accession == "NC_013965.1", "pHV2",
                               ifelse(genomic_accession == "NC_013964.1", "pHV3",
                                      ifelse(genomic_accession == "NC_013966.1", "pHV4", NA)))))
 })

gff_map <- within(gff_clean, {
    chr = ifelse(genomic_accession == "NC_013967.1", "Chromosome", 
                 ifelse(genomic_accession == "NC_013968.1", "pHV1", 
                        ifelse(genomic_accession == "NC_013965.1", "pHV2",
                               ifelse(genomic_accession == "NC_013964.1", "pHV3",
                                      ifelse(genomic_accession == "NC_013966.1", "pHV4", NA)))))
 })

hvo_gene_lengths <- data.frame("ncbi_id" = hvo_map$ncbi_id,
                               "length" = hvo_map$length) 
hvo_gene_lengths <- unique(hvo_gene_lengths)

## add Uri's updated arcog annotations
arcog_uri <- read.delim("annotations/arcog_uri_20191205.txt", header = TRUE)
hvo_map$wp <- hvo_map$ncbi_id
  hvo_map <- separate(data = hvo_map, col = wp, into = c("tmp", "wp"), sep = "cds_")
  hvo_map$tmp <- NULL
  hvo_map <- separate(data = hvo_map, col = wp, into = c("tmp1", "wp", "tmp2"), sep = "_")
  hvo_map$tmp1 <- NULL
  hvo_map$tmp2 <- NULL
  hvo_map$wp <- paste0('WP_', hvo_map$wp, sep = "")
  hvo_map <- merge(hvo_map, arcog_uri, by.x = "wp", by.y = "locus_tag", all = TRUE)

## rerun the counts table filter in previous chunk before rerunning the following
count_data2 <- count_data
count_data2$ncbi_id <- rownames(count_data2)
rownames(count_data2) <- NULL
annot_counts <- merge(hvo_map, count_data2, by = "ncbi_id")
write.table(annot_counts, file = "data/RSEM.txt",
            quote = FALSE, sep = "\t", row.names = FALSE) # I think this should be full_annot.txt

write.table(hvo_map, file = "annotations/hvo_rmap.txt",
            quote = FALSE, sep = "\t", row.names = FALSE)
# hvo_map <- read.delim("annotations/hvo_rmap.txt")
names(hvo_map)

rm(arcog_uri, cds_ids, gff_clean, gff_map, go_plus)
```

```{r error check ignore}
duplicated(hvo_map$ncbi_id)
any(duplicated(colnames(count_data)))
# any(duplicated(count_data$ncbi_id))
any(duplicated(rownames(count_data)))
length(unique(hvo_map$ncbi_id))
length(unique(hvo_map$wp))
length(unique(rownames(count_data_raw)))
length(unique(rownames(count_data)))
#keggmap <- dplyr::filter(keggmap,  !is.na(kegg_id))
#sub.keggmap <- aggregate(keggmap,by=list(keggmap$hvo),head,n=1)
n_occur <- data.frame(table(rownames(count_data)))
n_occur[n_occur$Freq > 1,]
count_data[rownames(count_data) %in% n_occur$Var1[n_occur$Freq > 1],]
# rownames(count_data) <- count_data$ncbi_id
# count_data$ncbi_id <- NULL
```


###\n Run differential expression analysis with noiseqbio
```{r noiseq analysis, results='hide', fig.keep='none', echo=FALSE}
dir.create("noiseq")
# Make noiseq object
dds = NOISeq::readData(data = count_data, factors = myfactors, length = hvo_gene_lengths)
save(dds, file ="noiseq/mydata_mate.RData")
#load(file ="noiseq/mydata_mate.RData")

# Run differential expression for data with biological replicates and construct volcano plots
for(s1 in time){
  for(s2 in time){
    if(s1!=s2){
      if(s1<s2){
        print(c(s1,s2))
        tmm <- noiseqbio(dds, norm = "tmm", k = 0.5, factor = "time", conditions = c(s1,s2), 
                             r = 50, filter = 1, nclust = 15, a0per= 0.9, random.seed = 12345, plot = TRUE, lc = 1)
        save(tmm, file = paste0("noiseq/mate_",s1,"vs",s2,"_tmm.RData"))
        tmm_r <- as.data.frame(tmm@results, row.names = NULL)
        write.table(tmm@results, paste0("noiseq/mate_",s1,"vs",s2,"_tmm.txt"), 
                    sep = "\t", row.names = TRUE, col.names = NA, quote = FALSE)
        tmm_r <- na.omit(tmm_r)
        tmm_r$P <- with(tmm_r, 1-prob)
        tmm_r$ncbi_id <- row.names(tmm_r)
        write.table(tmm_r, paste0("noiseq/mate_",s1,"vs",s2,"_tmm_filtered.txt"), 
                    sep = "\t", row.names = TRUE, col.names = NA, quote = FALSE)
        tmm_rm <- merge(tmm_r, hvo_map, by = "ncbi_id", all.x=TRUE)
        tmm_rm$EFFECTSIZE <- tmm_rm$log2FC
        write.table(tmm_rm, paste0("noiseq/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"),
                    sep = "\t", row.names = F, quote = FALSE)
}}}}
```
###\n Volcano plots
```{r volcano plots, out.width = "50%", echo=FALSE} 
# Volcano plots for publishing
for(s1 in time){
  for(s2 in time){
    if(s1!=s2){
      if(s1<s2){
        print(c(s1,s2))
        tmm_r <- read.delim(paste0("noiseq/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"), 
                            header=TRUE, row.names = NULL, com='', check.names=F)
        plot(tmm_r$log2FC, -log10(1-tmm_r$prob), pch=20, xlim=c(-7,7), ylim=c(0,16), main = paste(s1,"vs",s2), 
          xlab = "log2FC", ylab = "-log10(p-value)",
            col = ifelse(abs(tmm_r$log2FC)>5 & tmm_r$prob>0.999,"red",
              ifelse(abs(tmm_r$log2FC)>5,"orange",
                ifelse(tmm_r$prob>0.99,"green","black"))))
}}}}
```
###\n Interactive volcano plots labeled with *H. volcanii* identifier
```{r interactive volcano plots (hvo), echo=FALSE}
# Interactive volcano plots
for(s1 in time){
  for(s2 in time){
    if(s1!=s2){
      if(s1<s2){
        print(c(s1,s2))
        tmm_r <- read.delim(paste0("noiseq/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"), 
                            header=TRUE, row.names = NULL, com='', check.names=F)
        volc_obj <- volcanor(tmm_r, p = "P", effect_size = "EFFECTSIZE", snp = "kegg_id", gene = "locus_tag", 
                             annotation1 = "locus_tag", annotation2 = "protein")
        volc_plot <- volcanoly(volc_obj, effect_size_line = c(-5,5), effect_size_line_color = "orange", 
                  genomewideline = -log10(1e-3), genomewideline_color = "green", title = paste(s1,"vs",s2))
        print(volc_plot)
      }}}}
```
###\n Interactive volcano plotes labeled with kegg orthologous heirarchal categories
```{r interactive volcano plots (kegg), echo=FALSE}
# Interactive volcano plots labeled with kegg orthologies / pathways
for(s1 in time){
  for(s2 in time){
    if(s1!=s2){
      if(s1<s2){
        print(c(s1,s2))
        tmm_r <- read.delim(paste0("noiseq/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"), 
                            header=TRUE, row.names = NULL, com='', check.names=F)
        volc_obj <- volcanor(tmm_r, p = "P", effect_size = "EFFECTSIZE", snp = "keggidprot", 
                             gene = "kegg3", annotation1 = "kegg2", annotation2 = "kegg1")
        volc_plot <- volcanoly(volc_obj, effect_size_line = c(-5,5), effect_size_line_color = "orange", 
                  genomewideline = -log10(1e-3), genomewideline_color = "green", title = paste("kegg ",s1,"vs",s2))
        print(volc_plot)
      }}}}
```
###\n Principle component analysis
```{r pca, dev='pdf', fig.width=5, fig.height=5, out.width='600px', echo=FALSE}
# PCA plot - normally used to identify clustering patterns of biological and technical replicates; 
# ran on data without replicates to see if any patterns stood out based on nutrient source or time;
# differences seem to emerge between the two treatments over time
tpmMatrix = cpm(count_data_raw)
rnaseqMatrix = as.matrix(tpmMatrix)
rnaseqMatrix = round(rnaseqMatrix)
exp_study = DGEList(counts=rnaseqMatrix, group=factor(colnames(rnaseqMatrix)))
exp_study = calcNormFactors(exp_study, method = "TMM")
exp_study$samples$eff.lib.size = exp_study$samples$lib.size * exp_study$samples$norm.factors
tmm_table = cpm(exp_study)
mydata2 = readData(tmm_table, myfactors)
myPCA = dat(mydata2, type = "PCA")
explo.plot(myPCA, factor = "rep")
explo.plot(myPCA, factor = "time")
explo.plot(myPCA, factor = "samp")

pdf("plots/pca_rep.pdf")
explo.plot(myPCA, factor = "rep")
dev.off()
pdf("plots/pca_time.pdf")
explo.plot(myPCA, factor = "time")
dev.off()
pdf("plots/pca_samp.pdf")
explo.plot(myPCA, factor = "samp")
dev.off()
```

###\n Stats
```{r stats, eval=FALSE, echo=FALSE}
exp.pca <- prcomp(t(count_data), scale. = TRUE)
# summary(exp.pca)
# str(exp.pca)
# library(devtools)
# install_github("vqv/ggbiplot")
# library(ggbiplot)
# ggbiplot(exp.pca)
pc <- as.data.frame(exp.pca$x)
pc$heading <- rownames(pc)
pc_factors <- merge(pc, myfactors, by = "heading")
ggplot(pc_factors, aes(x = PC1, y = PC2, colour = time_de)) +
  geom_point(size = 3) +
  scale_color_manual(name = NULL, labels = c("shaking", "t00", "t02", "t04", "t08", "t24"), 
                  values=c("darkgrey", "#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.title = element_blank(),
        legend.position = c(0.1, 0.87), legend.background = element_blank(), legend.key.size = unit(0.35, "cm"))
ggsave("plots/pca.pdf", width = 4, height = 4, units = "in")

PC1_stats <- lm(PC1 ~ time_de, data = pc_factors)
summary(PC1_stats)
PC2_stats <- lm(PC2 ~ time_de, data = pc_factors)
summary(PC2_stats)
PC3_stats <- lm(PC3 ~ time_de, data = pc_factors)
summary(PC3_stats)

# fviz_pca_var(exp.pca,
#              col.var = "contrib", # Color by contributions to the PC
#              gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
#              repel = TRUE     # Avoid text overlapping
#              )

```

###\n Correlation tests on replicates - Pearson and Spearman
```{r correlations, warning=FALSE, echo=FALSE}
## correlation test for replicates

pear_corr <- cor(count_data, use="all.obs",method="pearson")
pear_corr
write.table(pear_corr, file = "tables/pear_corr.txt", sep = "\t", quote = FALSE)

spear_corr <- cor(count_data, use="all.obs",method="spearman")
spear_corr
write.table(spear_corr, file = "tables/spear_corr.txt", sep = "\t", quote = FALSE)

## add metadata to input table
tmm_ann <- as.data.frame(tmm_table)
tmm_ann$ncbi_id <- rownames(tmm_ann)
tmm_ann <- merge(tmm_ann, annomap, by = "ncbi_id", all.x=TRUE)
tmm_ann <- merge(tmm_ann, locus_tag, by = "hvo", all.x=TRUE)
tmm_ann <- merge(tmm_ann, keggmap, by = "old_locus_tag", all.x=TRUE)
names(tmm_ann)

## plot to examine relationship between points
for(s1 in rep1){
  for(s2 in rep2){
    if(substr(s1, start=2, stop=3)==substr(s2, start=2, stop=3)){
        print(c(s1,s2))
        lm_fit <- lm(paste(s1[[1]], "~", s1[[1]]), data=tmm_ann)
        summary(lm_fit)
        pred = data.frame(tmm_ann, predict(lm_fit, interval = 'prediction'))
        corr_plot <- ggplot(pred, aes_string(s1, s2)) + 
          geom_point() + #adds points
          theme_bw() + #removes grey background
          xlim(c(0,15000)) + ylim(c(0,15000)) + #reduces axes lengths to zoom in on majority of points
          theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), aspect.ratio=1) + #removes grid and makes square
          labs(x = (substr(s1, start=1, stop=3))) + labs(y = (substr(s2, start=1, stop=3))) + #renames labels from column headers
          geom_smooth(method='loess', level=0.95, aes(fill = 'confidence'), alpha = 0.5) + #adds regression line and 95% ci
          #geom_ribbon(aes(y = fit, ymin = lwr, ymax = upr, fill = 'prediction'), alpha = 0.2) +
          scale_fill_manual('Interval', values = c('grey', 'darkgreen')) 
        print(corr_plot)
        ggsave(paste0("plots/corr_plot_",s1,"vs",s2,".pdf"))
}}}
```

###\n Differential expression bar plots show higher expression in each timepoint compared to shaking as positive value and lower expression compared to shaking as negative value
```{r gene lists}
### This set of lists contains all original genes - hypotheticals/unused were removed in the list in the below chunk
lists <- list(
  `Glycosylation`=c("HVO_0727", "HVO_1524", "HVO_1527", "HVO_1528", "HVO_1529", "HVO_1530", "HVO_1531", "HVO_A0586", 
                    "HVO_1517", "HVO_1523A", "HVO_0798", "HVO_1526", "HVO_1522", "HVO_1523", "HVO_2072"),
  `Predicted Amyloid`=c("HVO_0066", "HVO_A0362", "HVO_A0410", "HVO_B0132", "HVO_B0138", "HVO_B0139", "HVO_B0140", 
                        "HVO_B0141", "HVO_B0222"),
  `Tubulin`=c("HVO_0581", "HVO_0717", "HVO_0745", "HVO_1113", "HVO_2013", "HVO_2068", "HVO_2204", "HVO_A0035"),
  `Chemotaxis`=c("HVO_0553", "HVO_0554", "HVO_1205", "HVO_1206", "HVO_1207", "HVO_1219", "HVO_1221", "HVO_1222", 
                 "HVO_1223", "HVO_1224", "HVO_1225", "HVO_2993"),
  `SNARE dedA Motif`=c("HVO_RS05125", "HVO_0934", "HVO_1969", "HVO_2517", "HVO_B0204"),
  `Quorum Sensing`=c("HVO_0204", "HVO_0205", "HVO_1811", "HVO_1812", "HVO_1917", "HVO_2337", "HVO_2338", "HVO_2339"),
  `Low-Salt Glycosylation`=c("HVO_2043", "HVO_2044", "HVO_2045", "HVO_2046", "HVO_2047", "HVO_2048", "HVO_2049", 
                             "HVO_2051", "HVO_2052", "HVO_2053", "HVO_2054", "HVO_2055", "HVO_2056", "HVO_2057", 
                             "HVO_2058", "HVO_2059", "HVO_2060", "HVO_2061", "HVO_2062", "HVO_2063", "HVO_2064"),
  `Molecular Parasites`=c("HVO_0858", "HVO_B0152", "HVO_C0080", "HVO_B0155", "HVO_A0209"),
  `MATE Family Transporters`=c("HVO_1524", "HVO_2055", "HVO_2342", "HVO_2652", "HVO_A0162", "HVO_A0197"),
  `Annotated Adhesins`=c("HVO_2397", "HVO_A0611"),
  `Bacterio-opsin Activator`=c("HVO_0318", "HVO_0492", "HVO_0513", "HVO_1117", "HVO_1190", "HVO_1356", "HVO_1357", 
                               "HVO_1952", "HVO_1953", "HVO_2835", "HVO_2836", "HVO_2894", "HVO_A0168", "HVO_A0169", 
                               "HVO_A0424", "HVO_A0495", "HVO_A0506", "HVO_B0020", "HVO_B0159", "HVO_B0272"),
  `Cell Division`=c("HVO_1725", "HVO_2042", "HVO_2293", "HVO_B0001", "HVO_A0001", "HVO_A0064", "HVO_A0257", "HVO_0220"),
  `Secretion System`=c("HVO_2166", "HVO_2167", "HVO_2168", "HVO_2169", "HVO_2170", "HVO_2171", "HVO_2172", "HVO_2173", 
                     "HVO_1477"),
  #`DNA Uptake`=c("HVO_1477"),
  `Type IV Secretion`=c("HVO_0421", "HVO_0620", "HVO_0748", "HVO_1160", "HVO_1217", "HVO_2278", "HVO_2385", "HVO_C0036"),
  `IS4 Family Transposase (set 1/3)`=c("HVO_B0152", "HVO_B0157", "HVO_B0234", "HVO_A0077", "HVO_A0081", "HVO_A0110", "HVO_A0125", 
                           "HVO_A0144", "HVO_A0151", "HVO_A0250", "HVO_A0260", "HVO_A0262", "HVO_A0367", "HVO_A0420", 
                           "HVO_A0423", "HVO_A0431", "HVO_A0433", "HVO_A0449", "HVO_0275", "HVO_0276", "HVO_0528"), 
  `IS4 Family Transposase (set 2/3)`=c("HVO_0922", "HVO_0926", "HVO_1519", "HVO_RS12025", "HVO_1525", "HVO_1782", "HVO_1835", 
                           "HVO_1845", "HVO_2051", "HVO_2054", "HVO_2252", "HVO_2331", "HVO_2466", "HVO_2817", 
                           "HVO_2820", "HVO_C0004", "HVO_C0021", "HVO_C0039", "HVO_C0053", "HVO_C0070", "HVO_C0071"), 
  `IS4 Family Transposase (set 3/3)`=c("HVO_A0007", "HVO_A0238", "HVO_A0355", "HVO_0924", "HVO_A0080", "HVO_A0456", "HVO_1106", 
                           "HVO_2830", "HVO_C0013", "HVO_A0116"),
  `IS5 Family Transposase`=c("HVO_A0117", "HVO_A0057", "HVO_A0240", "HVO_C0068", "HVO_0864", "HVO_B0249", "HVO_A0134"),
  `IS6 Family Transposase`=c("HVO_A0265", "HVO_C0065", "HVO_C0020A"),
  `ISH3 Family Transposase`=c("HVO_A0354", "HVO_A0408", "HVO_A0483", "HVO_2050", "HVO_A0353", "HVO_1837", "HVO_0263", 
                            "HVO_A0364", "HVO_A0445", "HVO_C0037", "HVO_A0349"),
  `IS630 Family Transposase`=c("HVO_A0100", "HVO_A0403", "HVO_C0078", "HVO_C0080"),
  `IS110 Family Transposase`=c("HVO_0556", "HVO_A0016", "HVO_A0128", "HVO_A0014", "HVO_A0279", "HVO_0278", "HVO_2821"),
  `IS1595 Family Transposase`=c("HVO_RS19485", "HVO_C0064", "HVO_A0062", "HVO_A0258", "HVO_RS19515"),
  `IS200/IS605 Family Transposase`=c("HVO_1150", "HVO_B0242"),
  `Other Transposase`=c("HVO_B0155", "HVO_B0152A", "HVO_B0241", "HVO_A0008", "HVO_A0018", "HVO_A0066", "HVO_A0309", 
                       "HVO_1151", "HVO_2075", "HVO_2434", "HVO_RS19915", "HVO_2090A"),
  `misc`=c("HVO_2380", "HVO_2700", "HVO_1327", "HVO_0915"),
  `new glyco`=c("HVO_2166", "HVO_2167", "HVO_2168", "HVO_2169", "HVO_2170", "HVO_2171", "HVO_2172"),
  `virus proteins`=c("HVO_0268", "HVO_0269", "HVO_0270", "HVO_0271", "HVO_0272", "HVO_0273", "HVO_0274", "HVO_0275", "HVO_0276",
                     "HVO_0279", "HVO_0258", "HVO_0260", "HVO_0261", "HVO_2278", "HVO_2824"),
  `RM System`=c("HVO_0794", "HVO_2269", "HVO_2270", "HVO_2271", "HVO_C0040", "HVO_A0006", "HVO_A0079", "HVO_A0237")
)
```

```{r bar plots, echo=FALSE, warning=FALSE, fig.width=7.5, fig.height=5}
#fig.width=7.5, fig.height=5
# Make differential expression table with fold changes to plot

## attempting to set up recursively to set object names with variable; still broken, but does make objects
## as it is below, each object is uniquely named, but contains all 5 dataframes
timesup <- c("00","02","04","08","24")
de <- sprintf("noiseq/mate_%svsp0_tmm.txt", timesup)
dfs <- lapply(de, read.delim, sep="")
#nam <- paste("de_", timesup, sep = "") 
#assign(nam, dfs[1:5])
#rm(de_00, de_02, de_04, de_08, de_24, dfs, de)

## stupidly inefficient - make recursive!!
de00 <- as.data.frame(dfs[1])
de02 <- as.data.frame(dfs[2])
de04 <- as.data.frame(dfs[3])
de08 <- as.data.frame(dfs[4])
de24 <- as.data.frame(dfs[5])
colnames(de00) <- c("de00_mean", "de00_p0_mean", "de00_theta", "de00_prob", "de00_log2fc", "de00_length")
colnames(de02) <- c("de02_mean", "de02_p0_mean", "de02_theta", "de02_prob", "de02_log2fc", "de02_length")
colnames(de04) <- c("de04_mean", "de04_p0_mean", "de04_theta", "de04_prob", "de04_log2fc", "de04_length")
colnames(de08) <- c("de08_mean", "de08_p0_mean", "de08_theta", "de08_prob", "de08_log2fc", "de08_length")
colnames(de24) <- c("de24_mean", "de24_p0_mean", "de24_theta", "de24_prob", "de24_log2fc", "de24_length")

to_shaking <- merge(de00, de02, by="row.names")
to_shaking <- merge(to_shaking, de04, by.x="Row.names", by.y="row.names")
to_shaking <- merge(to_shaking, de08, by.x="Row.names", by.y="row.names")
to_shaking <- merge(to_shaking, de24, by.x="Row.names", by.y="row.names")
logfc <- subset(to_shaking, select=c("Row.names", "de00_log2fc", "de02_log2fc", "de04_log2fc", "de08_log2fc", "de24_log2fc"))
sig <- subset(to_shaking, select=c("Row.names", "de00_prob", "de02_prob", "de04_prob", "de08_prob", "de24_prob"))



#rm(de00, de02, de04, de08, de24, dfs, de, timesup)
## annotate
## hvo_rmap_wgeneid_clean.txt has additional gene ids included added manually for specific genes examined here
## gene ids were identified using additional annotations from uniprot searches and kegg searches 
## that weren't originally outputted during annotation of the hvo genome
hvo_map_clean <- read.delim("annotations/hvo_rmap_wgeneid_clean.txt")
hvo_tmp <- unique(data.frame(hvo_map$ncbi_id, hvo_map$Gene_name))
# hvo_ultimate <- merge(hvo_map, hvo_map_clean)
oldnewgene <- NULL
  oldnewgene$gene <- hvo_map_clean$ncbi_id
  oldnewgene$old <- hvo_map_clean$old_locus_tag
  oldnewgene$new <- hvo_map_clean$locus_tag
  oldnewgene$prot <- hvo_map_clean$protein
  oldnewgene$geneid <- hvo_map_clean$gene_name
  oldnewgene$arcog <- hvo_map_clean$arcog
  oldnewgene$cog <- hvo_map_clean$cog
  oldnewgene <- as.data.frame(oldnewgene)
p0 <- merge(logfc, oldnewgene, by.x = "Row.names", by.y = "gene", all.x = TRUE)
p0 <- merge(p0, hvo_tmp, by.x = "Row.names", by.y = "hvo_map.ncbi_id", all = FALSE)
colnames(p0) <- c("Row.names", "de00","de02","de04","de08","de24", "old", "new", "prot", "geneid", "arcog", "cog", "gene_name")
pfull <- merge(to_shaking, oldnewgene, by.x = "Row.names", by.y = "gene", all = TRUE)
#rm(oldnewgene, to_shaking)
## lists of genes to plot together
## lists of genes to plot together
library(tidyr)
library(dplyr)
library(reshape2)
library(ggplot2)

### add significance stars
## make data.frame with a column containing x-labels and a column containing coordinates for the stars (bar height + 0.25)
## made sig object in above scripts
colnames(logfc) <- c("Row.names", "de00","de02","de04","de08","de24")
colnames(sig) <- c("Row.names", "de00","de02","de04","de08","de24")
logfcm <- melt(logfc, id = "Row.names")
colnames(logfcm) <- c("Row.names", "variable", "logfc")
sigm <- melt(sig, id = "Row.names")
colnames(sigm) <- c("Row.names", "variable", "prob")
sigmerge <- merge(sigm, logfcm)
sigmerge <- na.omit(sigmerge)
sighvo <- merge(sigmerge, oldnewgene, by.x = "Row.names", by.y = "gene", all.x = TRUE)
sig05 <- sighvo %>% filter(between(prob, 0.95, 0.99))
sig05$star <- NULL
sig01 <- filter(sighvo, prob >= 0.99)
## make coords for significant
coords <- NULL
coords$new <- sig05$new
coords$old <- sig05$old
coords$variable <- sig05$variable
coords$prot <- sig05$prot
coords$gene_name <- sig05$gene_name
coords$val <- ifelse(sig05$logfc>0, with(sig05, logfc + 0.05), with(sig05, logfc - 0.15))
coords <- as.data.frame(coords)
## make coords for highly significant
coords01 <- NULL
coords01$new <- sig01$new
coords01$old <- sig01$old
coords01$variable <- sig01$variable
coords01$prot <- sig01$prot
coords01$gene_name <- sig01$gene_name
coords01$val01 <- ifelse(sig01$logfc>0, with(sig01, logfc + 0.05), with(sig01, logfc - 0.15))
coords01 <- as.data.frame(coords01)

write.table(logfc, "data/logfc.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
write.table(sig, "data/sig.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

lists <- list(
  `Glycosylation`=c("HVO_0727", "HVO_1524", "HVO_1527", "HVO_1528", "HVO_1529", "HVO_1530", "HVO_1531", "HVO_A0586",
                    "HVO_1517", "HVO_1523A", "HVO_0798", "HVO_1526", "HVO_1522", "HVO_1523", "HVO_2072", "HVO_1734A"),
  `Predicted trans-membrane`=c("HVO_0088", "HVO_2517", "HVO_A0169", "HVO_2993", "HVO_0934", "HVO_1969", "HVO_0205",
                               "HVO_0727", "HVO_2342", "HVO_2652", "HVO_2055", "HVO_A0162",
                               "HVO_1524", "HVO_1530", "HVO_1524"),
  `DNA binding`=c("HVO_B0072", "HVO_B0148", "HVO_B0277", "HVO_A0598", "HVO_A0626", "HVO_0662", "HVO_0776", "HVO_1070",
                  "HVO_1141", "HVO_1929", "HVO_2002", "HVO_2067", "HVO_2151", "HVO_2869", "HVO_2953", "HVO_2970"),
  `Low-Salt Glycosylation`=c("HVO_2043", "HVO_2044", "HVO_2045", "HVO_2046", "HVO_2047", "HVO_2048", "HVO_2049",
                             "HVO_2051", "HVO_2052", "HVO_2053", "HVO_2054", "HVO_2055", "HVO_2056", "HVO_2057",
                             "HVO_2058", "HVO_2059", "HVO_2060", "HVO_2061", "HVO_2062", "HVO_2063", "HVO_2064"),
  `Quorum Sensing`=c("HVO_0204", "HVO_0205", "HVO_1811", "HVO_1812", "HVO_1917", "HVO_2337", "HVO_2338", "HVO_2339",
                     "HVO_1368", "HVO_2870"),
  `Secretion System`=c("HVO_2168", "HVO_0421", "HVO_1160", "HVO_2278"),
  `Chemotaxis`=c("HVO_2993", "HVO_1210", "HVO_1211", "HVO_A0607", "HVO_0420", "HVO_0554", "HVO_0555", "HVO_0969",
                 "HVO_1126", "HVO_1205", "HVO_1206", "HVO_1207", "HVO_1219", "HVO_1221", "HVO_1222", "HVO_1223", "HVO_1224",
                 "HVO_1225", "HVO_1484", "HVO_1779", "HVO_1999", "HVO_2214", "HVO_2220", "HVO_2462", "HVO_3005"),
  `flagella`=c("HVO_1033", "HVO_1034", "HVO_1159", "HVO_1160", "HVO_1200", "HVO_1203", "HVO_1213",
                "HVO_1214", "HVO_1215", "HVO_1216", "HVO_12174", "HVO_1218", "HVO_2385", "HVO_2386", "HVO_2993"),
  `EPS`=c("HVO_0915"),
  `dessication`=c("HVO_1184"),
  `stress`=c("HVO_B0095", "HVO_B0344", "HVO_A0043", "HVO_A0047", "HVO_A0086", "HVO_A0153", "HVO_A0175", "HVO_A0230", "HVO_A0344",
             "HVO_A0452", "HVO_A0496", "HVO_0033", "HVO_0102", "HVO_0176", "HVO_0401", "HVO_0428", "HVO_0536", "HVO_0612", "HVO_0904",
             "HVO_0931", "HVO_0932", "HVO_0938", "HVO_0948", "HVO_1087", "HVO_1184", "HVO_1198", "HVO_1267", "HVO_1319", "HVO_1481",
             "HVO_1589", "HVO_1592", "HVO_1778", "HVO_1820", "HVO_1822", "HVO_1823", "HVO_1853", "HVO_1874", "HVO_2156", "HVO_2234",
             "HVO_2239", "HVO_2337", "HVO_2366", "HVO_2484", "HVO_2499", "HVO_2500", "HVO_2534", "HVO_2615", "HVO_2684", "HVO_2788",
             "HVO_2899", "HVO_3009"),
  `division1`=c("HVO_B0001", "HVO_A0001", "HVO_A0064", "HVO_A0072", "HVO_A0257", "HVO_0001", "HVO_0120", "HVO_0194", "HVO_0225",
               "HVO_0228", "HVO_0392", "HVO_0581", "HVO_0595", "HVO_0634", "HVO_0717", "HVO_0745", "HVO_1113", "HVO_1327"),
  `division2`=c("HVO_1565", "HVO_1634", "HVO_1725", "HVO_1907", "HVO_1931", "HVO_2013", "HVO_2042", "HVO_2068", "HVO_2204",
               "HVO_2292", "HVO_2380", "HVO_2700", "HVO_C0001", "HVO_C0057", "HVO_A0035", "HVO_2293"),
  `Selfish genes`=c("HVO_0858", "HVO_B0152", "HVO_C0080", "HVO_B0155", "HVO_A0209"),
  `RM System`=c("HVO_0794", "HVO_2269", "HVO_2270", "HVO_2271", "HVO_C0040", "HVO_A0006", "HVO_A0079", "HVO_A0237"),
  `Tubulin`=c("HVO_0581", "HVO_0717", "HVO_0745", "HVO_1113", "HVO_2013", "HVO_2068", "HVO_2204", "HVO_A0035"),
  `Cell Division`=c("HVO_1725", "HVO_2042", "HVO_2293", "HVO_B0001", "HVO_A0001", "HVO_A0064", "HVO_A0257", "HVO_0220"),
  `SNARE dedA Motif`=c("HVO_RS05125", "HVO_0934", "HVO_1969", "HVO_2517", "HVO_B0204"),
  `excreted stuff`=c("HVO_2055"),
  `aglD`=c("HVO_0798", "HVO_1763", "HVO_1859"),
  `crispr`=c("HVO_A0205", "HVO_A0206", "HVO_A0207", "HVO_A0208","HVO_A0209", "HVO_A0210", "HVO_A0211", "HVO_A0212"),
  `all QS`=c("HVO_B0022", "HVO_B0023", "HVO_B0024", "HVO_B0079", "HVO_B0080", "HVO_B0081", "HVO_B0082", "HVO_B0089",
                         "HVO_B0090", "HVO_B0091", "HVO_B0092", "HVO_B0093", "HVO_B0184", "HVO_B0185", "HVO_B0186", "HVO_B0187",
                         "HVO_B0188", "HVO_B0189", "HVO_B0190", "HVO_B0219", "HVO_B0220", "HVO_B0221", "HVO_B0328", "HVO_B0331",
                         "HVO_A0174", "HVO_A0336", "HVO_A0337", "HVO_A0338", "HVO_A0339", "HVO_A0380", "HVO_A0381", "HVO_A0382",
                         "HVO_A0426A", "HVO_A0427", "HVO_A0514", "HVO_A0515", "HVO_A0516", "HVO_A0576", "HVO_A0579", "HVO_0058",
                         "HVO_0059", "HVO_0060", "HVO_0061", "HVO_0062", "HVO_0627", "HVO_0628", "HVO_0629", "HVO_0630", "HVO_0709",
                         "HVO_0900", "HVO_0903", "HVO_1194", "HVO_1195", "HVO_1368", "HVO_1917", "HVO_2122", "HVO_2123", "HVO_2124",
                         "HVO_2125", "HVO_2126", "HVO_2441", "HVO_2443", "HVO_2444", "HVO_2764", "HVO_2765", "HVO_2802", "HVO_C0072",
                         "HVO_C0073", "HVO_C0074", "HVO_C0075", "HVO_A0425", "HVO_A0426"),
  `andreas`=c("HVO_2166", "HVO_2167", "HVO_2168", "HVO_2169", "HVO_2170", "HVO_2171", "HVO_2172", "HVO_2173"),
  `Predicted Amyloid`=c("HVO_0066", "HVO_A0362", "HVO_A0410", "HVO_B0132", "HVO_B0138", "HVO_B0139", "HVO_B0140",
                        "HVO_B0141", "HVO_B0222"),
  `Secretion System`=c("HVO_2166", "HVO_2167", "HVO_2168", "HVO_2169", "HVO_2170", "HVO_2171", "HVO_2172", "HVO_2173"),
  `Type IV Secretion`=c("HVO_0421", "HVO_0620", "HVO_0748", "HVO_1160", "HVO_1217", "HVO_2278", "HVO_2385", "HVO_C0036"),
  `IS4 Family Transposase 1`=c("HVO_B0152", "HVO_B0157", "HVO_B0234", "HVO_A0077", "HVO_A0081", "HVO_A0110", "HVO_A0125",
                           "HVO_A0144", "HVO_A0151", "HVO_A0250", "HVO_A0260", "HVO_A0262", "HVO_A0367", "HVO_A0420",
                           "HVO_A0423", "HVO_A0431", "HVO_A0433", "HVO_A0449", "HVO_0275", "HVO_0276", "HVO_0528"),
  `IS4 Family Transposase 2`=c("HVO_0922", "HVO_0926", "HVO_1519", "HVO_RS12025", "HVO_1525", "HVO_1782", "HVO_1835",
                           "HVO_1845", "HVO_2051", "HVO_2054", "HVO_2252", "HVO_2331", "HVO_2466", "HVO_2817",
                           "HVO_2820", "HVO_C0004", "HVO_C0021", "HVO_C0039", "HVO_C0053", "HVO_C0070", "HVO_C0071"),
  `IS4 Family Transposase 3`=c("HVO_A0007", "HVO_A0238", "HVO_A0355", "HVO_0924", "HVO_A0080", "HVO_A0456", "HVO_1106",
                           "HVO_2830", "HVO_C0013", "HVO_A0116"),
  `IS5 Family Transposase`=c("HVO_A0117", "HVO_A0057", "HVO_A0240", "HVO_C0068", "HVO_0864", "HVO_B0249", "HVO_A0134"),
  `IS6 Family Transposase`=c("HVO_A0265", "HVO_C0065", "HVO_C0020A"),
  `ISH3 Family Transposase`=c("HVO_A0354", "HVO_A0408", "HVO_A0483", "HVO_2050", "HVO_A0353", "HVO_1837", "HVO_0263",
                            "HVO_A0364", "HVO_A0445", "HVO_C0037", "HVO_A0349"),
  `IS630 Family Transposase`=c("HVO_A0100", "HVO_A0403", "HVO_C0078", "HVO_C0080"),
  `IS110 Family Transposase`=c("HVO_0556", "HVO_A0016", "HVO_A0128", "HVO_A0014", "HVO_A0279", "HVO_0278", "HVO_2821"),
  `IS1595 Family Transposase`=c("HVO_RS19485", "HVO_C0064", "HVO_A0062", "HVO_A0258", "HVO_RS19515"),
  `IS200-IS605 Family Transposase`=c("HVO_1150", "HVO_B0242"),
  `Other Transposase`=c("HVO_B0155", "HVO_B0152A", "HVO_B0241", "HVO_A0008", "HVO_A0018", "HVO_A0066", "HVO_A0309",
                       "HVO_1151", "HVO_2075", "HVO_2434", "HVO_RS19915", "HVO_2090A")
)

## plot fold change with colorblind friendly palette
p <- htmltools::tagList()
for(i in 1:length(lists)){
  nm <- (names(lists)[i])
  print(nm)
  df_sub <- filter(p0, grepl(paste((lists[[i]]), collapse="|"), old))
  coords_sub <- filter(coords, grepl(paste((lists[[i]]), collapse="|"), old))
  coords01_sub <- filter(coords01, grepl(paste((lists[[i]]), collapse="|"), old))
  df_sub$Row.names <- NULL
  df_sub <- melt(df_sub, id.vars=c("old", "new", "prot", "geneid", "arcog", "cog", "gene_name"))
  df_sub1 <- merge(df_sub, coords_sub, all.x = TRUE)
  df_sub1 <- merge(df_sub1, coords01_sub, all.x = TRUE)
  p <- ggplot(data=df_sub1, aes(x= paste(old, prot, sep = '\n'), y=value, fill = variable)) +
    theme_bw() +
    geom_bar(stat="identity", position = position_dodge(width = NULL), width = 0.7) +
    geom_hline(yintercept = 0) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    labs(title = paste(nm,"genes"), 
         x = NULL, y = expression("log"["2"]*" fold change"), fill = "time vs shaking") +
    theme(axis.text.x = element_text(size  = 12, angle = 45, hjust = 1, vjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5), plot.margin = unit(c(0.5,0,0.5,4),"cm")) +
    scale_fill_manual(labels = c("t00", "t02", "t04", "t08", "t24"), 
                      values=c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")) +
    geom_text(aes(y = val), label = "*", position = position_dodge(width = 0.7)) +
    geom_text(aes(y = val01), label = "**", position = position_dodge(width = 0.7))
  print(p)
  nm2 <- gsub(" ", "_", nm)
  # ggsave(filename = paste("plots/",nm2,"_histogram.pdf", sep = ""))
  # ggsave(filename = paste("plots/",nm2,"_histogram.png", sep = ""))
}

write.table(p0, file = "data/p0.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
write.table(coords, file = "data/coords.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
write.table(coords01, file = "data/coords01.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

#ggplot_build(p)$data[[1]]
#ggplot_build(p)$data[[3]]

save(p0, file = "data/p0.RData")
save(coords, file = "data/coords.RData")
save(coords01, file = "data/coords01.RData")
```

```{r manuscript barplots}
load("data/p0.RData")
load("data/coords.RData")
load("data/coords01.RData")

coords$valt <- NULL
coords01$val01t <- NULL
coords$prot <- NULL
coords01$prot <- NULL

lists <- list(
  # `snare_deda`=c("HVO_RS05125", "HVO_0934", "HVO_1969", "HVO_2517", "HVO_B0204")
  # `cetz_ftsz`=c("HVO_0581", "HVO_0717", "HVO_0745", "HVO_1113", "HVO_2013", "HVO_2068", "HVO_2204", "HVO_A0035")
  # `cell_div`=c("HVO_1725", "HVO_2042", "HVO_2293", "HVO_B0001", "HVO_A0001", "HVO_A0064", "HVO_A0257", "HVO_0220"),
  # `rms`=c("HVO_0794", "HVO_2269", "HVO_2270", "HVO_2271", "HVO_C0040", "HVO_A0006", "HVO_A0079", "HVO_A0237")
  # `tivss`=c("HVO_0421", "HVO_0620", "HVO_0748", "HVO_1160", "HVO_1217", "HVO_2278", "HVO_2385", "HVO_C0036", "HVO_2281", "HVO_C0036")
  # `crispr_cas`=c("HVO_A0205", "HVO_A0206", "HVO_A0207", "HVO_A0208","HVO_A0209", "HVO_A0210", "HVO_A0211", "HVO_A0212")
  # `glyc`=c("HVO_0727", "HVO_1524", "HVO_1527", "HVO_1528", "HVO_1529", "HVO_1530", "HVO_1531",
  #          "HVO_1517", "HVO_1523A", "HVO_0798", "HVO_1526", "HVO_1522", "HVO_1523", "HVO_2072"),
  # `lsglyc`=c("HVO_2044", "HVO_2045", "HVO_2046", "HVO_2047", "HVO_2048", "HVO_2049",
  #            "HVO_2052", "HVO_2053", "HVO_2054", "HVO_2055", "HVO_2056", "HVO_2057",
  #            "HVO_2058", "HVO_2059", "HVO_2060", "HVO_2061", "HVO_2063", "HVO_2064")
  `nglyc_hyp`=c("HVO_2052", "HVO_A0198", "HVO_2391", "HVO_2267", "HVO_A0186", "HVO_A0214", "HVO_0702")
)

## relabel specific items in p0 based on corrections discussed in lab meetings
p0$prot <- as.character(p0$prot)
p0[2604, 9] <- "cell division protein FtsZ2"
p0[2514, 9] <- "cell division protein FtsZ1"
p0[2491, 9] <- "cell division protein CetZ2"
p0[2193, 9] <- "cell division protein CetZ3"
p0[1248, 9] <- "cell division protein CetZ5"
p0[1294, 9] <- "cell division protein CetZ6"
p0[3425, 9] <- "cell division protein CetZ1"
p0[635, 9] <- "cell division protein CetZ4"
p0[3406, 9] <- "type I RMS restriction subunit"
p0[3243, 9] <- "type I RMS methylation subunit"
p0[3342, 9] <- "type I RMS specificity subunit"
p0[3542, 9] <- "DNA cytosine methyltransferase"
p0[454, 9] <- "methyltransferase (hypothetical)"
p0[607, 9] <- "N-6 DNA methylase (hypothetical)"
p0[779, 9] <- "N-6 DNA methylase (hypothetical)"
p0[2113, 9] <- "type II/IV secretion system protein (hypothetical)"
p0[3556, 9] <- "type IV secretion system coupling protein DNA-binding domain (hypothetical)"
p0[1445, 9] <- "type IV secretion system TraD (conjugal transfer protein)"
p0[2729, 9] <- "type II/IV secretion system protein"
p0[1448, 9] <- "type IV secretory pathway, VirB4 component"
p0[2161, 9] <- "type II/IV secretion system protein"
p0[497, 9] <- "helicase Cas3"

## revision to add gene id to plot and add indicator for non-pathway genes
## all genes cross referenced in uniprot
# p0$geneid <- is.na(p0$geneid)
p0$geneid <- as.character(p0$geneid)
p0[411, 10] <- "GraD1" # uniprot
# p0[2505, 10] <- HVO_RS08195 no gene id
p0[2452, 10] <- "AglD"
p0[3309, 10] <- "AglJ"
p0[886, 10] <- "AglF"
p0[3379, 10] <- "AglI"
p0[887, 10] <- "AglG"
p0[3285, 10] <- "AglB"
p0[3285, 9] <- "oligosaccharyl transferase"
p0[888, 10] <- "AglM"
p0[3418, 10] <- "Csg"
p0[881, 10] <- "AglP"
p0[882, 10] <- "AglQ"
p0[882, 9] <- "agl cluster protein"
p0[883, 10] <- "AglE"
p0[884, 10] <- "AglR"
p0[884, 9] <- "agl cluster protein"
p0[885, 10] <- "AglS"
p0[3352, 10] <- "EndH"
p0[3352, 9] <- "endonuclease I"
p0[1276, 10] <- "Agl7"
p0[1278, 10] <- "Agl9"
p0[1279, 10] <- "Agl10"
p0[1283, 10] <- "Agl13"
p0[1284, 10] <- "Agl11"
p0[3272, 10] <- "Agl14"
p0[1286, 10] <- "Agl12"
p0[1287, 10] <- "Agl8"
p0[1288, 10] <- "Agl6"
p0[1281, 10] <- "Agl5"
p0[1282, 10] <- "Agl15"
# p0[, 10] <- ""

## revisions to protein name for uniformity 
p0[411, 9] <- "Sugar nucleotidyltransferase" # uniprot

p0$prot <- as.factor(p0$prot)

for(i in 1:length(lists)){
  nm <- (names(lists)[i])
  print(nm)
  df_sub <- filter(p0, grepl(paste((lists[[i]]), collapse="|"), old))
  coords_sub <- filter(coords, grepl(paste((lists[[i]]), collapse="|"), old))
  coords01_sub <- filter(coords01, grepl(paste((lists[[i]]), collapse="|"), old))
  df_sub$Row.names <- NULL
  df_sub <- melt(df_sub, id.vars=c("old", "new", "prot", "geneid", "arcog", "cog", "gene_name"))
  df_sub1 <- merge(df_sub, coords_sub, all.x = TRUE)
  df_sub1 <- merge(df_sub1, coords01_sub, all.x = TRUE)
  p <- ggplot(data=df_sub1, aes(x= paste(paste(new, old, sep = ' / '), 
                                         # paste(prot, geneid, sep = ' '), # for glyc
                                         str_wrap(paste(prot, sep = ' '), width = 35), # for rm, ss, and crispr
                                         # paste(prot, sep = ' '), # for snare and cetz
                                          sep = '\n'), # multiline x-axis
                                y=value, fill = variable)) +
    theme_bw() +
    geom_bar(stat="identity", position = position_dodge(width = NULL), width = 0.7) +
    geom_hline(yintercept = 0) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    labs(title = paste(nm,"genes"), 
         x = NULL, y = expression("log"["2"]*" fold change"), fill = "time vs shaking") +
    theme(text = element_text(size=12), axis.text.x = 
            element_text(size  = 12, angle = 45, hjust = 1, vjust = 1)) +
    theme(plot.title = element_blank()) +
    # theme(plot.margin = unit(c(0.5,0.5,0.5,2), "cm")) + ## use this for most plot margins
    theme(plot.margin = unit(c(0.5,1,0,2), "cm")) + ## use this for cowplot for rm
    # theme(plot.margin = unit(c(0.5,1,0,5), "cm")) + ## use this for cowplot for glyc

    scale_fill_manual(labels = c("t00", "t02", "t04", "t08", "t24"), 
                      values = c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")) +
    theme(legend.background = element_rect(fill = "transparent")) +
    theme(legend.position = c(0.9, 0.85)) +
    theme(legend.title = element_text(size = 10), legend.text = element_text(size = 10)) +
    theme(legend.key.size = unit(0.07,"in")) +
    geom_text(aes(y = val), label = "*", position = position_dodge(width = 0.7)) +
    geom_text(aes(y = val01), label = "**", position = position_dodge(width = 0.7))
  print(p)
  nm2 <- gsub(" ", "_", nm)
  assign(paste("plot", nm2, sep = "_"), p)
 # ggsave(filename = paste("plots/bar_",nm2,"_histogram_revision.pdf", sep = ""),
         # height = 6, width = 8, units = "in" ## for snare
        # height = 6, width = 7, units = "in" ## for tivss
         # height = 8, width = 10, units = "in" ## for ftsz/cetz system
        # )
  # ggsave(filename = paste("exp_mating/gene_plots/",nm2,"_histogram.png", sep = ""))
}

### for final plots, edit script as needed - slightly modified for each
```

```{r long plot biofilm, fig.width=6.5, fig.height=7}
### gene names manually adjusted in p0 table
p0 <- read.delim("data/p0_newgene.txt")
coords <- read.delim("data/coords.txt")
coords01 <- read.delim("data/coords01.txt")

coords$val <- ifelse(sig05$logfc>0, with(sig05, logfc + 0.05), with(sig05, logfc - 0.05))
coords01$val01 <- ifelse(sig01$logfc>0, with(sig01, logfc + 0.05), with(sig01, logfc - 0.05))
# coords$star <- ifelse(sig05$logfc>0, "*", "*")
coords$star <- "*"
coords01$star <- "**"

p0$prot <- as.character(p0$prot)
p0[1544, 9] <- "secretion system protein PilC3"
p0[2055, 9] <- "transducer protein Htr15b"
p0[2054, 9] <- "transducer protein Htr15a"
p0[2053, 9] <- "transducer protein Htr15a"
p0[751, 9] <- "peptidase A24 PibD"
p0[2059, 9] <- "methyl-accepting chemotaxis protein Htr7"
p0[553, 9] <- "chemotaxis response regulator protein CheB"
p0[554, 9] <- "chemotaxis response regulator protein CheB"
p0[555, 9] <- "response regulator CheY"
p0[556, 9] <- "response regulator CheY"
p0[2058, 9] <- "methylâˆ’accepting chemotaxis protein Htr39"
p0[2435, 9] <- "ATPase MoxR3"
p0[842, 9] <- "type II secretion system protein PilB3"
p0[843, 9] <- "secretion system protein PilB4"
p0[3075, 9] <- "Flp pilus assembly complex ATPase component TadA"
p0[2356, 9] <- "archaeosortase A ArtA"

p0[834, 9] <- "archaellar protein ArlCE"
p0[835, 9] <-  "archaellin cluster protein ArlD1"
p0[836, 9] <-  "archaellin cluster protein ArlD2"
p0[837, 9] <-  "archaellin cluster protein ArlF"
p0[838, 9] <-  "archaellin cluster protein ArlG"
p0[839, 9] <- "archaellin ArlA2"
p0[840, 9] <- "ATPase ArlH"
p0[2418, 9] <- "archaellin ArlA1"


p0[2469, 9] <- "pilin PilA2"
p0[2470, 9] <- "pilin PilA3"
p0[2471, 9] <- "pilin PilA4"
p0[2472, 9] <- "pilin PilA5"
p0[2473, 9] <- "pilin PilA6"
p0[2814, 9] <- "type IV pilin"
p0$prot <- as.factor(p0$prot)

secretion_system <- c("HVO_2168", "HVO_0421", "HVO_2278", "HVO_0915")
chemotaxis <- c("HVO_2993", "HVO_0420", "HVO_0555", "HVO_0969",
               "HVO_1207", "HVO_1219", "HVO_1222", "HVO_1224",
               "HVO_1225", "HVO_1779", "HVO_1999", "HVO_2214", "HVO_3005")
flagella <- c("HVO_1033", "HVO_1034", "HVO_1160", "HVO_1200", "HVO_1203", 
              "HVO_1210", "HVO_1211", "HVO_1213", "HVO_1214", "HVO_1215", 
              "HVO_1216", "HVO_2062", "HVO_2450", "HVO_2451", 
              "HVO_A0632", "HVO_A0633", "HVO_2704")
# removed: "HVO_1217", "HVO_1218", 
# other_flagella <- c("HVO_2062", "HVO_2450", "HVO_2451", "HVO_A0632", "HVO_A0633", "HVO_2704") # added to flagella

df <- stack(mget(c('secretion_system', 'chemotaxis', 'flagella')))
df <- plyr::rename(df, c("values"="old", "ind"="category"))

df_sub <- dplyr::distinct(merge(df, p0, all.x = TRUE, by = "old"))
df_sub$Row.names <- NULL
df_sub <- reshape::melt(df_sub, id.vars=c("old", "category", "new", "prot", "gene_name", "arcog"))
coords[ ,c('new', 'prot')] <- list(NULL)
coords01[ ,c('new', 'prot')] <- list(NULL)
df_sub1 <- left_join(df_sub, coords, by = c("old", "variable"))
df_sub1 <- left_join(df_sub1, coords01, by = c("old", "variable"))
df_sub1 <- unite(df_sub1, "star", c(star.x, star.y), na.rm = TRUE, sep = '')

ggplot(data=arrange(df_sub1, category), aes(x = old, y=value, fill = variable, group = variable)) +
    theme_bw() +
    geom_bar(stat="identity", position = position_dodge2(width = NULL, reverse=TRUE, padding = 0), width = 0.7) +
    geom_hline(yintercept = 0) +
    annotate("text", label = "Secretion Systems", x = 1, y = 2.7, size = 4) +
    geom_vline(xintercept = 4.5, col='grey47', lwd=1) +
    annotate("text", label = "Chemotaxis", x = 5, y = 2.7, size = 4) +
    geom_vline(xintercept = 17.5, col='grey47', lwd=1) +
    annotate("text", label = "Archaella and Pili", x = 18, y = 2.7, size = 4) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    theme(axis.text.x = element_text(size = 8, hjust = 1, vjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5), plot.margin = unit(c(0.5,0,0.5,1),"cm")) +
    scale_fill_manual(labels = c("t00", "t02", "t04", "t08", "t24"), 
                      values=c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")) +
    geom_text(aes(label = star), stat = "identity",
              position = position_dodge2(width = 0.6, reverse = TRUE)) +
    aes(x = fct_inorder(paste(paste(new, old, sep = ' / '), prot, sep='\n'), category)) +
    labs(x = NULL, y = expression("log"["2"]*" fold change"), fill = "time vs shaking") +
    theme(legend.position = c(0.87, 0.9)) +
    coord_flip()
# longplot
ggsave("plots/fig3_community_revision.pdf", height = 11, width = 8)
ggsave("plots/fig3_community_revision.eps", height = 11, width = 8)
## added significance stars by hand due to misalignment with above script
```

```{r long plot insertion seqs, fig.width=5, fig.height=10}
p0 <- read.delim("data/p0_newgene.txt")
coords <- read.delim("data/coords.txt")
coords01 <- read.delim("data/coords01.txt")

coords$valt <- ifelse(coords$val>0, coords$val, coords$val + 0.1)
coords01$val01t <- ifelse(coords01$val01>0, coords01$val01, coords01$val01 + 0.1)


`Other Transposase`=c("HVO_B0155", "HVO_B0152A", "HVO_B0241", "HVO_A0008", "HVO_A0018", "HVO_A0066", "HVO_A0309",
                     "HVO_1151", "HVO_2075", "HVO_2434", "HVO_RS19915", "HVO_2090A") # 12
`IS200-IS605 Family Transposase`=c("HVO_1150", "HVO_B0242") # 2
`IS1595 Family Transposase`=c("HVO_RS19485", "HVO_C0064", "HVO_A0062", "HVO_A0258", "HVO_RS19515") # 4
`IS110 Family Transposase`=c("HVO_0556", "HVO_A0016", "HVO_A0128", "HVO_A0014", "HVO_A0279", "HVO_0278", "HVO_2821") # 3
`IS630 Family Transposase`=c("HVO_A0100", "HVO_A0403", "HVO_C0078", "HVO_C0080") # 4
`ISH3 Family Transposase`=c("HVO_A0354", "HVO_A0408", "HVO_A0483", "HVO_2050", "HVO_A0353", "HVO_1837", "HVO_0263",
                          "HVO_A0364", "HVO_A0445", "HVO_C0037", "HVO_A0349") # 10
`IS6 Family Transposase`=c("HVO_A0265", "HVO_C0065", "HVO_C0020A") # 3
`IS5 Family Transposase`=c("HVO_A0117", "HVO_A0057", "HVO_A0240", "HVO_C0068", "HVO_0864", "HVO_B0249", "HVO_A0134") # 6
`IS4 Family Transposase`=c("HVO_B0152", "HVO_B0157", "HVO_B0234", "HVO_A0077", "HVO_A0081", "HVO_A0110", "HVO_A0125",
                         "HVO_A0144", "HVO_A0151", "HVO_A0250", "HVO_A0260", "HVO_A0262", "HVO_A0367", "HVO_A0420",
                         "HVO_A0423", "HVO_A0431", "HVO_A0433", "HVO_A0449", "HVO_0275", "HVO_0276", "HVO_0528",
                         "HVO_0922", "HVO_0926", "HVO_1519", "HVO_RS12025", "HVO_1525", "HVO_1782", "HVO_1835",
                         "HVO_1845", "HVO_2051", "HVO_2054", "HVO_2252", "HVO_2331", "HVO_2466", "HVO_2817",
                         "HVO_2820", "HVO_C0004", "HVO_C0021", "HVO_C0039", "HVO_C0053", "HVO_C0070", "HVO_C0071",
                         "HVO_A0007", "HVO_A0238", "HVO_A0355", "HVO_0924", "HVO_A0080", "HVO_A0456", "HVO_1106",
                         "HVO_2830", "HVO_C0013", "HVO_A0116")

df <- stack(mget(c('Other Transposase', 'IS200-IS605 Family Transposase', 'IS1595 Family Transposase', 
                   'IS110 Family Transposase', 'IS630 Family Transposase', 'ISH3 Family Transposase',
                   'IS6 Family Transposase', 'IS5 Family Transposase', 'IS4 Family Transposase')))
df <- plyr::rename(df, c("values"="old", "ind"="category"))
df_sub <- dplyr::distinct(merge(df, p0, all.x = TRUE, by = "old"))
df_sub$Row.names <- NULL
df_sub <- reshape::melt(df_sub, id.vars=c("old", "category", "new", "prot", "gene_name", "arcog"))
coords_sub <- dplyr::distinct(merge(df, coords, all.x = TRUE, by = "old"))
coords01_sub <- dplyr::distinct(merge(df, coords01, all.x = TRUE, by = "old"))
df_sub1 <- merge(df_sub, coords_sub, all.x = TRUE)
df_sub1 <- merge(df_sub1, coords01_sub, all.x = TRUE)
df_sub1 <- df_sub1[!is.na(df_sub1$value), ]

is_table <- dplyr::distinct(merge(df, p0, all.x = TRUE, by = "old"))
  is_table <- dplyr::distinct(merge(is_table, sig, all.x = TRUE, by = "Row.names"))
  is_table[,c(1,11:12)] <- NULL
  is_table <- is_table[,c(8,1:2,9,3:7,10:14)]
  is_table %>% rename_at(vars(ends_with(".x")), funs(str_replace(., ".x", ""))) -> is_table
  is_table %>% rename_at(vars(ends_with(".y")), funs(str_replace(., ".y", "_probability"))) -> is_table
write.table(is_table, "tables/table_s5_insertion_seqs.txt", quote = FALSE, sep = "\t", row.names = FALSE)

ggplot(data=arrange(df_sub1, category), aes(x = old, y=value, fill = variable, group = variable)) +
# ggplot(data=arrange(df_sub1,category), aes(x = paste(new, prot, sep='\n'), y=value, fill = variable)) +
    theme_bw() +
    geom_bar(stat="identity", position = position_dodge2(width = NULL, reverse=TRUE, padding = 0), width = 0.7) +
    # geom_text(position = position_dodge(width = 1), aes(x=category, y=0)) +
    geom_hline(yintercept = 0) +
    # geom_text(aes(0, 4.5, label = "other", vjust = -0.7), color = "grey47", size = 3) +
    geom_vline(xintercept = 12.5, col='grey47', lwd=0.5) + 
  geom_text(aes(12.5, 5, label = "other", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_vline(xintercept = 14.5, col='grey47', lwd=0.5) + 
  geom_text(aes(14.5, 5, label = "IS200-IS605 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3)+
    geom_vline(xintercept = 18.5, col='grey47', lwd=0.5) + 
  geom_text(aes(18.5, 5, label = "IS1595 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_vline(xintercept = 21.5, col='grey47', lwd=0.5) + 
  geom_text(aes(21.5, 5, label = "IS110 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_vline(xintercept = 25.5, col='grey47', lwd=0.5) + 
  geom_text(aes(25.5, 5, label = "IS630 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_vline(xintercept = 35.5, col='grey47', lwd=0.5) + 
  geom_text(aes(35.5, 5, label = "ISH3 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_vline(xintercept = 38.5, col='grey47', lwd=0.5) + 
  geom_text(aes(38.5, 5, label = "IS6 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_vline(xintercept = 44.5, col='grey47', lwd=0.5) + 
  geom_text(aes(44.5, 5, label = "IS5 Family", vjust = 1.2, hjust = 1), color = "grey47", size = 3) +
    geom_text(aes(81.5, 5, label = "IS4 Family", vjust = 1.1, hjust = 1), color = "grey47", size = 3) +
    ylim(-5, 5) +
    theme(legend.position = c(0.12,0.94)) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    theme(axis.text.x = element_text(size = 8, hjust = 1, vjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5), plot.margin = unit(c(0.5,0,0.5,1),"cm")) +
    scale_fill_manual(labels = c("t00", "t02", "t04", "t08", "t24"), 
                      values=c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")) +
    # geom_text(aes(y = valt), label = "*", position = position_dodge(width = 0.7)) +
    # geom_text(aes(y = val01t), label = "**", position = position_dodge(width = 0.7)) +
    aes(x = fct_inorder(paste(new, old, sep = ' / '), category)) +
    labs(x = NULL, y = expression("log"["2"]*" fold change"), fill = "time vs shaking") +
    coord_flip()
  # facet_wrap(~category, nrow = 1, scales = "free_x")
# longplot
ggsave("plots/figS3_transposases_revision.pdf", height = 15, width = 8)
ggsave("plots/figS3_transposases_revision.eps", height = 15, width = 8)
```

```{r violin plot}
### violin plot for transposases
## this chunk should run independently of rest of the script
library(beanplot)
library(purrr)
library(NOISeq)
library(ggplot2)
library(wesanderson)
library(grid)

count_data_raw <- read.table("data/mating.mat", 
                             header=T, com='', check.names=F)
  count_data <- round(count_data_raw, 0)
  count_data <- count_data_raw[rowSums(count_data_raw) > 10,]
  p0tmm <- tmm(count_data, long = 1000, lc = 0, k = 0, refColumn = 1, logratioTrim = 0.3, 
               sumTrim = 0.05, doWeighting = TRUE, Acutoff = -1e+10)
  p0tmm <- as.data.frame(p0tmm)
  p0tmm$Row.names <- rownames(p0tmm)

p0 <- read.delim("data/p0.txt")
p0mod <- p0
  p0mod$de00 <- NULL
  p0mod$de02 <- NULL
  p0mod$de04 <- NULL
  p0mod$de08 <- NULL
  p0mod$de24 <- NULL
p0mer <- merge(p0mod, p0tmm, by = "Row.names")

transposase <- c("HVO_B0152", "HVO_B0157", "HVO_B0234", "HVO_A0077", "HVO_A0081", 
                 "HVO_A0110", "HVO_A0125", "HVO_A0144", "HVO_A0151", "HVO_A0250", 
                 "HVO_A0260", "HVO_A0262", "HVO_A0367", "HVO_A0420", "HVO_A0423", 
                 "HVO_A0431", "HVO_A0433", "HVO_A0449", "HVO_0275", "HVO_0276", 
                 "HVO_0528", "HVO_0922", "HVO_0926", "HVO_1519", "HVO_RS12025", 
                 "HVO_1525", "HVO_1782", "HVO_1835", "HVO_1845", "HVO_2051", "HVO_2054", 
                 "HVO_2252", "HVO_2331", "HVO_2466", "HVO_2817", "HVO_2820", "HVO_C0004", 
                 "HVO_C0021", "HVO_C0039", "HVO_C0053", "HVO_C0070", "HVO_C0071", 
                 "HVO_A0007", "HVO_A0238", "HVO_A0355", "HVO_0924", "HVO_A0080", 
                 "HVO_A0456", "HVO_1106", "HVO_2830", "HVO_C0013", "HVO_A0116", 
                 "HVO_A0117", "HVO_A0057", "HVO_A0240", "HVO_C0068", "HVO_0864", 
                 "HVO_B0249", "HVO_A0134", "HVO_A0265", "HVO_C0065", "HVO_C0020A", 
                 "HVO_A0354", "HVO_A0408", "HVO_A0483", "HVO_2050", "HVO_A0353", 
                 "HVO_1837", "HVO_0263", "HVO_A0364", "HVO_A0445", "HVO_C0037", 
                 "HVO_A0349","HVO_A0100", "HVO_A0403", "HVO_C0078", "HVO_C0080",
                 "HVO_0556", "HVO_A0016", "HVO_A0128", "HVO_A0014", "HVO_A0279", 
                 "HVO_0278", "HVO_2821","HVO_RS19485", "HVO_C0064", "HVO_A0062", 
                 "HVO_A0258", "HVO_RS19515","HVO_1150", "HVO_B0242","HVO_B0155", 
                 "HVO_B0152A", "HVO_B0241", "HVO_A0008", "HVO_A0018", "HVO_A0066", 
                 "HVO_A0309", "HVO_1151", "HVO_2075", "HVO_2434", "HVO_RS19915", "HVO_2090A")

## tlist is a list of transposases by family
tlist <- read.delim("annotations/tlist.txt", header = FALSE)
  tlist <- melt(tlist, id.vars="V1")
  tlist$variable <- NULL
p0sub <- subset(p0mer, old %in% transposase)
  p0sub$Row.names <- NULL
  p0sub$shaking <- rowMeans(p0sub[c('ap0_quant', 'bp0_quant')], na.rm=TRUE)
  p0sub$t00 <- rowMeans(p0sub[c('a00_quant', 'b00_quant')], na.rm=TRUE)
  p0sub$t02 <- rowMeans(p0sub[c('a02_quant', 'b02_quant')], na.rm=TRUE)
  p0sub$t04 <- rowMeans(p0sub[c('a04_quant', 'b04_quant')], na.rm=TRUE)
  p0sub$t08 <- rowMeans(p0sub[c('a08_quant', 'b08_quant')], na.rm=TRUE)
  p0sub$t24 <- rowMeans(p0sub[c('a24_quant', 'b24_quant')], na.rm=TRUE)
  p0sub <- merge(p0sub, tlist, by.x = "old", by.y = "value", all.x = TRUE)
  p0sub <- p0sub[, -c(4:19)]
  p0sub <- melt(p0sub, id.vars=c("old", "new", "prot", "V1"))

#beanplot(value~V1, data = p0sub, log = "", col = "bisque", method = "jitter")

viol <- ggplot(data = p0sub, aes(x = variable, y = value)) +
  geom_violin() +
  geom_jitter(height = 0, width = 0.2, aes(color = V1), alpha = 0.5) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  scale_color_manual(values = wes_palette(n = 9, name = "Darjeeling1", type = "continuous"), name = "") +
  labs(x = NULL, y = expression("log"["10"]*" TMM values")) +
  scale_y_continuous(trans='log10') +
  theme(legend.position = c(0.82, 0.18)) +
  theme(legend.key.height=unit(0.3,"line")) +
  theme(legend.key.width=unit(1,"line")) +
  theme(legend.background = element_rect(fill = "transparent")) +
  theme(plot.margin = unit(c(0,1,0,0.3),"cm")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
ggsave("plots/violin_transposase.pdf", height = 4, width = 7, units = "in", dpi = 600)
viol
```

```{r manhattan prep}
# manhattan plot
#### this chunk should run independently of the rest of the script
library(manhattanly)
library(reshape)
library(dplyr)

### files made in earlier chunks that were saved for read-in
hvo_map <- read.delim("annotations/hvo_rmap.txt")
logfc <- read.delim("data/logfc.txt")
sig <- read.delim("data/sig.txt")

## use hvo_map to start the merge dataframe for the manhattan plot input (see HapMap)
man_map <- NULL
man_map$CHR <- hvo_map$genomic_accession
man_map$CHR2 <- hvo_map$genomic_accession
man_map$BP <- hvo_map$start
man_map$SNP <- hvo_map$locus_tag
man_map$tmp <- hvo_map$ncbi_id
man_map$gene <- hvo_map$protein
man_map <- as.data.frame(man_map)

## use the significance and fold change data from each timepoint compared to a shaking control to make input
sig_melt <- melt(sig, id.vars = "Row.names")
  sig_melt <- dplyr::rename(sig_melt, prob = value, timepoint = variable)
  sig_melt$P <- with(sig_melt, 1 - prob)
logfc_melt <- melt(logfc, id.vars = "Row.names")
  logfc_melt <- dplyr::rename(logfc_melt, EFFECTSIZE = value, timepoint = variable)

man_in <- merge(logfc_melt, sig_melt, by = c("Row.names", "timepoint"))
man_in <- merge(man_in, man_map, by.x = "Row.names", by.y = "tmp")

## rename the chromosome/plasmids and clean up the dataframe
man_in$CHR <- gsub('NC_013967.1', 1, man_in$CHR)
  man_in$CHR <- gsub('NC_013968.1', 10, man_in$CHR)
  man_in$CHR <- gsub('NC_013965.1', 20, man_in$CHR)
  man_in$CHR <- gsub('NC_013964.1', 30, man_in$CHR)
  man_in$CHR <- gsub('NC_013966.1', 40, man_in$CHR)
  man_in$CHR <- as.numeric(man_in$CHR)
man_in$CHR2 <- gsub('NC_013967.1', 'Chromosome', man_in$CHR2)
  man_in$CHR2 <- gsub('NC_013968.1', 'pHV1', man_in$CHR2)
  man_in$CHR2 <- gsub('NC_013965.1', 'pHV2', man_in$CHR2)
  man_in$CHR2 <- gsub('NC_013964.1', 'pHV3', man_in$CHR2)
  man_in$CHR2 <- gsub('NC_013966.1', 'pHV4', man_in$CHR2)
man_in$Row.names <- NULL
man_in <- man_in[c("CHR", "CHR2", "BP", "P", "SNP", "EFFECTSIZE", "timepoint", "prob", "gene")]

manobj <- manhattanr(man_in, chr = "CHR", bp = "BP", p = "P", snp = "SNP", 
                     annotation1 = "timepoint", annotation2 = "EFFECTSIZE", gene = "gene", logp = TRUE)
intplot <- manhattanly(manobj, point_size = 5, labelChr = c("Chr", "pHV1", "pHV3", "pHV4"), xlab = "Haloferax volcanii",
            highlight = c("HVO_RS00735", "HVO_RS00740", "HVO_RS00765", "HVO_RS00775", "HVO_RS01150", "HVO_RS01185", "HVO_RS01190", "HVO_RS01220", "HVO_RS01895", "HVO_RS01900", "HVO_RS01930", "HVO_RS01935", "HVO_RS01940", "HVO_RS02130", "HVO_RS02155", "HVO_RS02175", "HVO_RS02220", "HVO_RS02235", "HVO_RS02240", "HVO_RS02380", "HVO_RS02395", "HVO_RS02400", "HVO_RS02430", "HVO_RS02445", "HVO_RS02500", "HVO_RS02535", "HVO_RS02910", "HVO_RS02920", "HVO_RS02960", "HVO_RS02990", "HVO_RS02995", "HVO_RS03005", "HVO_RS03015", "HVO_RS03080", "HVO_RS03235", "HVO_RS03435", "HVO_RS03440", "HVO_RS03480", "HVO_RS03490", "HVO_RS03670", "HVO_RS03720", "HVO_RS03730", "HVO_RS03765", "HVO_RS03770", "HVO_RS03825", "HVO_RS03835", "HVO_RS03870", "HVO_RS04000", "HVO_RS05975", "HVO_RS06035", "HVO_RS06040", "HVO_RS06050", "HVO_RS07250", "HVO_RS07375", "HVO_RS08840", "HVO_RS09115", "HVO_RS09120", "HVO_RS09130", "HVO_RS10010", "HVO_RS10230", "HVO_RS10235", "HVO_RS12020", "HVO_RS12025", "HVO_RS12030", "HVO_RS13270", "HVO_RS13510", "HVO_RS13560", "HVO_RS14565", "HVO_RS14570", "HVO_RS14575", "HVO_RS14680", "HVO_RS15525", "HVO_RS15925", "HVO_RS16425", "HVO_RS16590", "HVO_RS18325", "HVO_RS18335", "HVO_RS18340", "HVO_RS18385", "HVO_RS19305", "HVO_RS19350", "HVO_RS19370", "HVO_RS19430", "HVO_RS19435", "HVO_RS19485", "HVO_RS19490", "HVO_RS19515", "HVO_RS19535", "HVO_RS19540", "HVO_RS19550", "HVO_RS19560", "HVO_RS19565", "HVO_RS19605", "HVO_RS19775", "HVO_RS19810", "HVO_RS19915", "HVO_RS19920", "HVO_RS19935", "HVO_RS19970", "HVO_RS20195", "HVO_RS20245", "HVO_RS20355", "HVO_RS20415"))
htmlwidgets::saveWidget(widget = intplot, file = "manhattan_transposases.html", selfcontained = TRUE)

intplotc <- manhattanly(manobj, point_size = 5, labelChr = c("Chr", "pHV1", "pHV3", "pHV4"), xlab = "Haloferax volcanii",
            highlight = c("HVO_RS02755", "HVO_RS02760", "HVO_RS02765", "HVO_RS02770", "HVO_RS02780", "HVO_RS02785", "HVO_RS02790"),
            highlight_color = "blue")
htmlwidgets::saveWidget(widget = intplotc, file = "manhattan_crisprs.html", selfcontained = TRUE)

hvos <- c("HVO_RS08530", "HVO_RS09750", "HVO_RS12015", "HVO_RS12035", "HVO_RS12055", "HVO_RS13185", "HVO_RS13625",
                          "HVO_RS14535", "HVO_RS15825", "HVO_RS20150", "HVO_RS20160",
                          "HVO_RS04470", "HVO_RS08195", "HVO_RS08530", "HVO_RS12015", "HVO_RS12035", "HVO_RS12040", "HVO_RS12045", 
                          "HVO_RS12050", "HVO_RS12055", "HVO_RS14660", "HVO_RS20145", "HVO_RS20150", "HVO_RS20155", "HVO_RS20160",
                          "HVO_RS20165",
                          "HVO_RS14530", "HVO_RS14535", "HVO_RS14540", "HVO_RS14545", "HVO_RS14550", "HVO_RS14555", "HVO_RS14560", 
                          "HVO_RS14570", "HVO_RS14580", "HVO_RS14585", "HVO_RS14595", "HVO_RS14600", "HVO_RS14605", "HVO_RS14610",
                          "HVO_RS14615", "HVO_RS14620", "HVO_RS20230", "HVO_RS20235", "HVO_RS20240")
#count(man_in, vars = hvos)
library(expss)
for(i in hvos){
tmp <- count_if(i, man_in)
print(paste(i, "=", tmp))
}

intplot_agl <- manhattanly(manobj, point_size = 5, labelChr = c("Chr", "pHV1", "pHV3", "pHV4"), xlab = "Haloferax volcanii",
            highlight = c("HVO_RS09750", "HVO_RS13185", "HVO_RS13625", "HVO_RS15825",
                          "HVO_RS04470", "HVO_RS08195", "HVO_RS08530", "HVO_RS12015", "HVO_RS12035", "HVO_RS12040", "HVO_RS12045", 
                          "HVO_RS12050", "HVO_RS12055", "HVO_RS14660", "HVO_RS20145", "HVO_RS20150", "HVO_RS20155", "HVO_RS20160",
                          "HVO_RS20165",
                          "HVO_RS14530", "HVO_RS14535", "HVO_RS14540", "HVO_RS14545", "HVO_RS14550", "HVO_RS14555", "HVO_RS14560", 
                          "HVO_RS14570", "HVO_RS14580", "HVO_RS14585", "HVO_RS14595", "HVO_RS14600", "HVO_RS14605", "HVO_RS14610",
                          "HVO_RS14615", "HVO_RS14620", "HVO_RS20230", "HVO_RS20235", "HVO_RS20240"),
            highlight_color = c("blue", "blue", "blue", "blue",
                                "blue", "blue", "blue", "blue",
                                "blue", "blue", "blue", "blue",
                                "blue", "blue", "blue", "blue",
                                "blue", "blue", "blue", "blue",
                                "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", 
                                "green", "green", "green", "green",
                                "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", 
                                "green", "green", "green", "green",
                                "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", 
                                "green", "green", "green", "green",
                                "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", 
                                "green", "green", "green", "green",
                                "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", 
                                "green", "green", "green", "green",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange",
                                "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange", "orange"))
htmlwidgets::saveWidget(widget = intplot_agl, file = "manhattan_agl.html", selfcontained = TRUE)
```

```{r manhattan}
### from https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_Manhattan_Plots_in_R
library(lattice)
manhattan.plot <- function(chr, pos, pvalue, 
	sig.level=NA, annotate=NULL, ann.default=list(),
	should.thin=T, thin.pos.places=2, thin.logp.places=2, 
	xlab="Chromosome", ylab=expression(-log[10](p-value)),
	col=c("gray","darkgray"), panel.extra=NULL, pch=20, cex=0.8,...) {

	if (length(chr)==0) stop("chromosome vector is empty")
	if (length(pos)==0) stop("position vector is empty")
	if (length(pvalue)==0) stop("pvalue vector is empty")

	#make sure we have an ordered factor
	if(!is.ordered(chr)) {
		chr <- ordered(chr)
	} else {
		chr <- chr[,drop=T]
	}

	#make sure positions are in kbp
	if (any(pos>1e6)) pos<-pos/1e6;

	#calculate absolute genomic position
	#from relative chromosomal positions
	posmin <- tapply(pos,chr, min);
	posmax <- tapply(pos,chr, max);
	posshift <- head(c(0,cumsum(posmax)),-1);
	names(posshift) <- levels(chr)
	genpos <- pos + posshift[chr];
	getGenPos<-function(cchr, cpos) {
		p<-posshift[as.character(cchr)]+cpos
		return(p)
	}

	#parse annotations
	grp <- NULL
	ann.settings <- list()
	label.default<-list(x="peak",y="peak",adj=NULL, pos=3, offset=0.5, 
		col=NULL, fontface=NULL, fontsize=NULL, show=F)
	parse.label<-function(rawval, groupname) {
		r<-list(text=groupname)
		if(is.logical(rawval)) {
			if(!rawval) {r$show <- F}
		} else if (is.character(rawval) || is.expression(rawval)) {
			if(nchar(rawval)>=1) {
				r$text <- rawval
			}
		} else if (is.list(rawval)) {
			r <- modifyList(r, rawval)
		}
		return(r)
	}

	if(!is.null(annotate)) {
		if (is.list(annotate)) {
			grp <- annotate[[1]]
		} else {
			grp <- annotate
		} 
		if (!is.factor(grp)) {
			grp <- factor(grp)
		}
	} else {
		grp <- factor(rep(1, times=length(pvalue)))
	}
  
	ann.settings<-vector("list", length(levels(grp)))
	ann.settings[[1]]<-list(pch=pch, col=col, cex=cex, fill=col, label=label.default)

	if (length(ann.settings)>1) { 
		lcols<-trellis.par.get("superpose.symbol")$col 
		lfills<-trellis.par.get("superpose.symbol")$fill
		for(i in 2:length(levels(grp))) {
			ann.settings[[i]]<-list(pch=pch, 
				col=lcols[(i-2) %% length(lcols) +1 ], 
				fill=lfills[(i-2) %% length(lfills) +1 ], 
				cex=cex, label=label.default);
			ann.settings[[i]]$label$show <- T
		}
		names(ann.settings)<-levels(grp)
	}
	for(i in 1:length(ann.settings)) {
		if (i>1) {ann.settings[[i]] <- modifyList(ann.settings[[i]], ann.default)}
		ann.settings[[i]]$label <- modifyList(ann.settings[[i]]$label, 
			parse.label(ann.settings[[i]]$label, levels(grp)[i]))
	}
	if(is.list(annotate) && length(annotate)>1) {
		user.cols <- 2:length(annotate)
		ann.cols <- c()
		if(!is.null(names(annotate[-1])) && all(names(annotate[-1])!="")) {
			ann.cols<-match(names(annotate)[-1], names(ann.settings))
		} else {
			ann.cols<-user.cols-1
		}
		for(i in seq_along(user.cols)) {
			if(!is.null(annotate[[user.cols[i]]]$label)) {
				annotate[[user.cols[i]]]$label<-parse.label(annotate[[user.cols[i]]]$label, 
					levels(grp)[ann.cols[i]])
			}
			ann.settings[[ann.cols[i]]]<-modifyList(ann.settings[[ann.cols[i]]], 
				annotate[[user.cols[i]]])
		}
	}
 	rm(annotate)

	#reduce number of points plotted
	if(should.thin) {
		thinned <- unique(data.frame(
			logp=round(-log10(pvalue),thin.logp.places), 
			pos=round(genpos,thin.pos.places), 
			chr=chr,
			grp=grp)
		)
		logp <- thinned$logp
		genpos <- thinned$pos
		chr <- thinned$chr
		grp <- thinned$grp
		rm(thinned)
	} else {
		logp <- -log10(pvalue)
	}
	rm(pos, pvalue)
	gc()

	#custom axis to print chromosome names
	axis.chr <- function(side,...) {
		if(side=="bottom") {
			panel.axis(side=side, outside=T,
				at=((posmax+posmin)/2+posshift),
				labels=levels(chr), 
				ticks=F, rot=0,
				check.overlap=F
			)
		} else if (side=="top" || side=="right") {
			panel.axis(side=side, draw.labels=F, ticks=F);
		}
		else {
			axis.default(side=side,...);
		}
	 }

	#make sure the y-lim covers the range (plus a bit more to look nice)
	prepanel.chr<-function(x,y,...) { 
		A<-list();
		maxy<-ceiling(max(y, ifelse(!is.na(sig.level), -log10(sig.level), 0)))+.5;
		A$ylim=c(0,maxy);
		A;
	}

	xyplot(logp~genpos, chr=chr, groups=grp,
		axis=axis.chr, ann.settings=ann.settings, 
#		auto.key=TRUE,
		prepanel=prepanel.chr, scales=list(axs="i"),
		panel=function(x, y, ..., getgenpos) {
			if(!is.na(sig.level)) {
				#add significance line (if requested)
				panel.abline(h=-log10(sig.level), lty=2);
			}
			panel.superpose(x, y, ..., getgenpos=getgenpos);
			if(!is.null(panel.extra)) {
				panel.extra(x,y, getgenpos, ...)
			}
		},
		panel.groups = function(x,y,..., subscripts, group.number) {
			A<-list(...)
			#allow for different annotation settings
			gs <- ann.settings[[group.number]]
			A$col.symbol <- gs$col[(as.numeric(chr[subscripts])-1) %% length(gs$col) + 1]    
			A$cex <- gs$cex[(as.numeric(chr[subscripts])-1) %% length(gs$cex) + 1]
			A$alpha <- 0.4
			A$pch <- gs$pch[(as.numeric(chr[subscripts])-1) %% length(gs$pch) + 1]
			A$fill <- gs$fill[(as.numeric(chr[subscripts])-1) %% length(gs$fill) + 1]
			A$x <- x
			A$y <- y
			do.call("panel.xyplot", A)
			#draw labels (if requested)
			if(gs$label$show) {
				gt<-gs$label
				names(gt)[which(names(gt)=="text")]<-"labels"
				gt$show<-NULL
				if(is.character(gt$x) | is.character(gt$y)) {
					peak = which.max(y)
					center = mean(range(x))
					if (is.character(gt$x)) {
						if(gt$x=="peak") {gt$x<-x[peak]}
						if(gt$x=="center") {gt$x<-center}
					}
					if (is.character(gt$y)) {
						if(gt$y=="peak") {gt$y<-y[peak]}
					}
				}
				if(is.list(gt$x)) {
					gt$x<-A$getgenpos(gt$x[[1]],gt$x[[2]])
				}
				do.call("panel.text", gt)
			}
		},
		xlab=xlab, ylab=ylab, 
		panel.extra=panel.extra, getgenpos=getGenPos, ...
	);
}

transposases <- list("HVO_RS00735", "HVO_RS00740", "HVO_RS00765", "HVO_RS00775", "HVO_RS01150", 
                     "HVO_RS01185", "HVO_RS01190", "HVO_RS01220", "HVO_RS01895", "HVO_RS01900", 
                     "HVO_RS01930", "HVO_RS01935", "HVO_RS01940", "HVO_RS02130", "HVO_RS02155", 
                     "HVO_RS02175", "HVO_RS02220", "HVO_RS02235", "HVO_RS02240", "HVO_RS02380", 
                     "HVO_RS02395", "HVO_RS02400", "HVO_RS02430", "HVO_RS02445", "HVO_RS02500", 
                     "HVO_RS02535", "HVO_RS02910", "HVO_RS02920", "HVO_RS02960", "HVO_RS02990", 
                     "HVO_RS02995", "HVO_RS03005", "HVO_RS03015", "HVO_RS03080", "HVO_RS03235", 
                     "HVO_RS03435", "HVO_RS03440", "HVO_RS03480", "HVO_RS03490", "HVO_RS03670", 
                     "HVO_RS03720", "HVO_RS03730", "HVO_RS03765", "HVO_RS03770", "HVO_RS03825", 
                     "HVO_RS03835", "HVO_RS03870", "HVO_RS04000", "HVO_RS05975", "HVO_RS06035", 
                     "HVO_RS06040", "HVO_RS06050", "HVO_RS07250", "HVO_RS07375", "HVO_RS08840", 
                     "HVO_RS09115", "HVO_RS09120", "HVO_RS09130", "HVO_RS10010", "HVO_RS10230", 
                     "HVO_RS10235", "HVO_RS12020", "HVO_RS12025", "HVO_RS12030", "HVO_RS13270", 
                     "HVO_RS13510", "HVO_RS13560", "HVO_RS14565", "HVO_RS14570", "HVO_RS14575", 
                     "HVO_RS14680", "HVO_RS15525", "HVO_RS15925", "HVO_RS16425", "HVO_RS16590", 
                     "HVO_RS18325", "HVO_RS18335", "HVO_RS18340", "HVO_RS18385", "HVO_RS19305", 
                     "HVO_RS19350", "HVO_RS19370", "HVO_RS19430", "HVO_RS19435", "HVO_RS19485", 
                     "HVO_RS19490", "HVO_RS19515", "HVO_RS19535", "HVO_RS19540", "HVO_RS19550", 
                     "HVO_RS19560", "HVO_RS19565", "HVO_RS19605", "HVO_RS19775", "HVO_RS19810", 
                     "HVO_RS19915", "HVO_RS19920", "HVO_RS19935", "HVO_RS19970", "HVO_RS20195", 
                     "HVO_RS20245", "HVO_RS20355", "HVO_RS20415")

ann<-rep(1, length(man_in$SNP))
ann[with(man_in, timepoint=="de00" & SNP %in% transposases)]<-2
ann[with(man_in, timepoint=="de02" & SNP %in% transposases)]<-3
ann[with(man_in, timepoint=="de04" & SNP %in% transposases)]<-4
ann[with(man_in, timepoint=="de08" & SNP %in% transposases)]<-5
ann[with(man_in, timepoint=="de24" & SNP %in% transposases)]<-6
ann<-factor(ann, levels=1:6)

manplot <- manhattan.plot(man_in$CHR2, man_in$BP, man_in$P, xlab = "",
              annotate = list(ann, "2" = list(label = " ", col = "#000000"),
                  "3" = list(label = " ", col = "#E69F00"),
                  "4" = list(label = " ", col = "#009E73"),
                  "5" = list(label = " ", col = "#0072B2"),
                  "6" = list(label = " ", col = "#D55E00")),
              key = list(border=T, padding.text=2, 
            	corner=c(.95, .95), text=list(lab=c("t00","t02","t04","t08","t24")), 
            	points=list(col=c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00"), pch=20)))
manplot

# pdf("plots/manplot_transposase.pdf", height = 4, width = 6)
# print(manplot)
# dev.off

### repeat for crispr/cas genes
crispr <- list("HVO_RS02755", "HVO_RS02760", "HVO_RS02765", "HVO_RS02770", "HVO_RS02780", "HVO_RS02785", "HVO_RS02790")

annc<-rep(1, length(man_in$SNP))
annc[with(man_in, timepoint=="de00" & SNP %in% crispr)]<-2
annc[with(man_in, timepoint=="de02" & SNP %in% crispr)]<-3
annc[with(man_in, timepoint=="de04" & SNP %in% crispr)]<-4
annc[with(man_in, timepoint=="de08" & SNP %in% crispr)]<-5
annc[with(man_in, timepoint=="de24" & SNP %in% crispr)]<-6
annc<-factor(annc, levels=1:6)

manplotc <- manhattan.plot(man_in$CHR2, man_in$BP, man_in$P, xlab = "",
              annotate = list(annc, "2" = list(label = " ", col = "#000000"),
                              "3" = list(label = " ", col = "#E69F00"),
                              "4" = list(label = " ", col = "#009E73"),
                              "5" = list(label = " ", col = "#0072B2"),
                              "6" = list(label = " ", col = "#D55E00")),
              key = list(border=T, padding.text=2, 
            	corner=c(.95, .95), text=list(lab=c("t00","t02","t04","t08","t24")), 
            	points=list(col=c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00"), pch=20)))
manplotc

# pdf("plots/manplot_crispr.pdf", height = 4, width = 6)
# print(manplotc)
# dev.off
# c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")
```

```{r cowplot}
## each plot made in individual chunks
manplot
plot_rms
viol
plot_grid(manplot, viol, plot_rms, labels = c('A', 'B', 'C'), label_size = 14, 
          ncol = 1, nrow = 3, rel_heights = c(0.9, 0.9, 1.2))
# ggsave("plots/figure6_selfish_genes.pdf", height = 15, width = 8, units = "in")
ggsave("plots/figure6_selfish_genes_revision.pdf", height = 15, width = 8, units = "in")

manplotc
plot_crispr_cas
plot_grid(manplotc, plot_crispr_cas, labels = c('A', 'B'), label_size = 14, 
          ncol = 1, nrow = 2, rel_heights = c(0.9, 1.1))
# ggsave("plots/figureS4_crispr.pdf", height = 12, width = 10, units = "in")
ggsave("plots/figureS4_crispr_revision.pdf", height = 12, width = 10, units = "in")

## ggsave won't include alpha so plots are blank and ggplot device = cairo_ps/cairo_eps not working
cairo_ps(file = "plots/figure6_selfish_genes_revision.eps", onefile = TRUE, 
         fallback_resolution = 300, height = 15, width = 8)
plot_grid(manplot, viol, plot_rms, labels = c('A', 'B', 'C'), label_size = 14, 
          ncol = 1, nrow = 3, rel_heights = c(0.9, 0.9, 1.2))
dev.off()

cairo_ps(file = "plots/figureS4_crispr_revision.eps", onefile = TRUE, 
         fallback_resolution = 300, height = 12, width = 10)
plot_grid(manplotc, plot_crispr_cas, labels = c('A', 'B'), label_size = 14, 
          ncol = 1, nrow = 2, rel_heights = c(0.9, 1.1))
dev.off()


plot_glyc
plot_lsglyc
plot_grid(plot_glyc, plot_lsglyc, labels = c('A', 'B'), label_size = 18, ncol = 1, nrow = 2)
ggsave("plots/figure_2_glycosylation_revision.pdf", height = 18, width = 14, units = "in")
ggsave("plots/figure_2_glycosylation_revision.eps", height = 18, width = 14, units = "in")
```

```{r top and bottom de}
load("data/p0.RData")
load("data/coords.RData")
load("data/coords01.RData")

#head(logfc)
#for(i in c("de00","de02","de04","de08","de24")) {
top_n(p0, 10, de00) %>% arrange(-de00) %>%
  write.table(file = "tables/de00_top10.txt", quote = FALSE, sep = "\t", row.names = FALSE) 
top_n(p0, 10, de02) %>% arrange(-de02) %>%
  write.table(file = "tables/de02_top10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
top_n(p0, 10, de04) %>% arrange(-de04) %>%
  write.table(file = "tables/de04_top10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
top_n(p0, 10, de08) %>% arrange(-de08) %>%
  write.table(file = "tables/de08_top10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
top_n(p0, 10, de24) %>% arrange(-de24) %>%
  write.table(file = "tables/de24_top10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
#max(logfc$de00, na.rm = TRUE)

top_n(p0, -10, de00) %>% arrange(-de00) %>%
  write.table(file = "tables/de00_bottom10.txt", quote = FALSE, sep = "\t", row.names = FALSE) 
top_n(p0, -10, de02) %>% arrange(-de02) %>%
  write.table(file = "tables/de02_bottom10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
top_n(p0, -10, de04) %>% arrange(-de04) %>%
  write.table(file = "tables/de04_bottom10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
top_n(p0, -10, de08) %>% arrange(-de08) %>%
  write.table(file = "tables/de08_bottom10.txt", quote = FALSE, sep = "\t", row.names = FALSE)
top_n(p0, -10, de24) %>% arrange(-de24) %>%
  write.table(file = "tables/de24_bottom10.txt", quote = FALSE, sep = "\t", row.names = FALSE)

#top 20
# lists <- list(
#   `de00 top 20`=c("HVO_2375", "HVO_RS19695", "HVO_RS19895", "HVO_B0212", "HVO_0280A", "HVO_A0184", "HVO_2672", "HVO_0925", "HVO_A0319", "HVO_RS04575", "HVO_C0086", "HVO_A0399", "HVO_0365", "HVO_RS11620", "HVO_RS19980", "HVO_1644", "HVO_0956", "HVO_0267", "HVO_0852", "HVO_RS03310"),
#   `de02 top 20`=c("HVO_2550", "HVO_2545", "HVO_0925", "HVO_RS19695", "HVO_A0004", "HVO_B0212", "HVO_2544", "HVO_2553", "HVO_2562", "HVO_C0086", "HVO_0280A", "HVO_0852", "HVO_2649", "HVO_A0399", "HVO_2546", "HVO_2543", "HVO_2549", "HVO_1981", "HVO_2547", "HVO_2551"),
#   `de04 top 20`=c("HVO_0925", "HVO_RS19695", "HVO_A0004", "HVO_A0399", "HVO_B0212", "HVO_2649", "HVO_1644", "HVO_A0272", "HVO_RS19895", "HVO_0280A", "HVO_1705", "HVO_0852", "HVO_B0129", "HVO_0133", "HVO_A0059", "HVO_0257", "HVO_2550", "HVO_C0086", "HVO_A0098", "HVO_2545"),
#   `de08 top 20`=c("HVO_B0212", "HVO_RS19895", "HVO_1705", "HVO_A0399", "HVO_B0198", "HVO_B0047", "HVO_B0041", "HVO_0383", "HVO_1644", "HVO_A0052", "HVO_2550", "HVO_0280A", "HVO_A0272", "HVO_A0059", "HVO_B0129", "HVO_A0319", "HVO_2980", "HVO_B0224", "HVO_A0268", "HVO_0925"),
#   `de24 top 20`=c("HVO_RS19695", "HVO_0665", "HVO_A0399", "HVO_A0611", "HVO_B0041", "HVO_B0212", "HVO_A0319", "HVO_RS19895", "HVO_A0029", "HVO_1705", "HVO_0925", "HVO_B0047", "HVO_B0198", "HVO_B0046", "HVO_2038", "HVO_B0042", "HVO_B0129", "HVO_2980", "HVO_RS04575", "HVO_0257")
# )

#bottom 20
# lists <- list(
#   `de00 bottom 20`=c("HVO_2733", "HVO_0493", "HVO_1889", "HVO_A0497", "HVO_C0078", "HVO_0177", "HVO_0200", "HVO_2718", "HVO_0947", "HVO_A0359", "HVO_0368", "HVO_A0434", "HVO_1525", "HVO_2741", "HVO_0378", "HVO_1805", "HVO_RS11000", "HVO_0171", "HVO_0373", "HVO_A0408A"),
#   `de02 bottom 20`=c("HVO_0200", "HVO_1757", "HVO_1671", "HVO_0855", "HVO_2179", "HVO_2243", "HVO_0947", "HVO_RS11000", "HVO_0873", "HVO_RS20250", "HVO_1795", "HVO_0515", "HVO_2238", "HVO_2569", "HVO_0171", "HVO_1801", "HVO_2184", "HVO_A0454", "HVO_2987", "HVO_A0408A"),
#   `de04 bottom 20`=c("HVO_0096", "HVO_2241", "HVO_1537", "HVO_0378", "HVO_2179", "HVO_0200", "HVO_0515", "HVO_2243", "HVO_2372", "HVO_0932", "HVO_RS11000", "HVO_2973", "HVO_0373", "HVO_1889", "HVO_0947", "HVO_RS19670", "HVO_0931", "HVO_A0592", "HVO_0171", "HVO_2121"),
#   `de08 bottom 20`=c("HVO_0763", "HVO_0249", "HVO_0489", "HVO_1489", "HVO_2037", "HVO_0488", "HVO_A0142", "HVO_A0548", "HVO_1643", "HVO_1525", "HVO_RS20250", "HVO_1795", "HVO_B0362", "HVO_2570", "HVO_0947", "HVO_2920", "HVO_0171", "HVO_2238", "HVO_2184", "HVO_A0408A"),
#   `de24 bottom 20`=c("HVO_A0548", "HVO_0355", "HVO_0772", "HVO_2973", "HVO_1753", "HVO_RS20250", "HVO_1751", "HVO_2282", "HVO_0502", "HVO_1108", "HVO_1924", "HVO_2178", "HVO_1278", "HVO_2920", "HVO_RS20340", "HVO_2569", "HVO_0947", "HVO_RS14225", "HVO_A0408A", "HVO_1325")
#   )

#top/bottom 10
lists <- list(
  `t00 top 10`=c("HVO_2375", "HVO_RS19695", "HVO_RS19895", "HVO_B0212", "HVO_0280A", "HVO_A0184", "HVO_2672", "HVO_0925", "HVO_A0319", "HVO_RS04575"),
  `t02 top 10`=c("HVO_2550", "HVO_2545", "HVO_0925", "HVO_RS19695", "HVO_A0004", "HVO_B0212", "HVO_2544", "HVO_2553", "HVO_2562", "HVO_C0086"),
  `t04 top 10`=c("HVO_0925", "HVO_RS19695", "HVO_A0004", "HVO_A0399", "HVO_B0212", "HVO_2649", "HVO_1644", "HVO_A0272", "HVO_RS19895", "HVO_0280A"),
  `t08 top 10`=c("HVO_B0212", "HVO_RS19895", "HVO_1705", "HVO_A0399", "HVO_B0198", "HVO_B0047", "HVO_B0041", "HVO_0383", "HVO_1644", "HVO_A0052"),
  `t24 top 10`=c("HVO_RS19695", "HVO_0665", "HVO_A0399", "HVO_A0611", "HVO_B0041", "HVO_B0212", "HVO_A0319", "HVO_RS19895", "HVO_A0029", "HVO_1705"),
  `t00 bottom 10`=c("HVO_2733", "HVO_0493", "HVO_1889", "HVO_A0497", "HVO_C0078", "HVO_0177", "HVO_0200", "HVO_2718", "HVO_0947", "HVO_A0359"),
  `t02 bottom 10`=c("HVO_0200", "HVO_1757", "HVO_1671", "HVO_0855", "HVO_2179", "HVO_2243", "HVO_0947", "HVO_RS11000", "HVO_0873", "HVO_RS20250"),
  `t04 bottom 10`=c("HVO_0096", "HVO_2241", "HVO_1537", "HVO_0378", "HVO_2179", "HVO_0200", "HVO_0515", "HVO_2243", "HVO_2372", "HVO_0932"),
  `t08 bottom 10`=c("HVO_0763", "HVO_0249", "HVO_0489", "HVO_1489", "HVO_2037", "HVO_0488", "HVO_A0142", "HVO_A0548", "HVO_1643", "HVO_1525"),
  `t24 bottom 10`=c("HVO_A0548", "HVO_0355", "HVO_0772", "HVO_2973", "HVO_1753", "HVO_RS20250", "HVO_1751", "HVO_2282", "HVO_0502", "HVO_1108")
  )

# p <- htmltools::tagList()
for(i in 1:length(lists)){
  nm <- (names(lists)[i])
  print(nm)
  df_sub <- filter(p0, grepl(paste((lists[[i]]), collapse="|"), old))
  coords_sub <- filter(coords, grepl(paste((lists[[i]]), collapse="|"), old))
  coords01_sub <- filter(coords01, grepl(paste((lists[[i]]), collapse="|"), old))
  df_sub$Row.names <- NULL
  df_sub <- melt(df_sub, id.vars=c("old", "new", "prot", "gene_name", "geneid", "cog", "arcog"))
  df_sub1 <- merge(df_sub, coords_sub, all.x = TRUE)
  df_sub1 <- merge(df_sub1, coords01_sub, all.x = TRUE)
  p <- ggplot(data=df_sub1, aes(x= paste(paste(new, old, sep = " / "), str_wrap(prot, width = 30), sep='\n'), 
                                y=value, fill = variable)) +
    theme_bw() +
    geom_bar(stat="identity", position = position_dodge(width = NULL), width = 0.7) +
    geom_hline(yintercept = 0) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    labs(title = paste(nm,"genes"), 
         x = NULL, y = expression("log"["2"]*" fold change"), fill = "time vs shaking") +
    theme(axis.text.x = element_text(size  = 16, angle = 45, hjust = 1, vjust = 1)) +
    theme(plot.title = element_text(hjust = 0.5), plot.margin = unit(c(0.5,0.5,0.5,3),"cm")) +
    scale_fill_manual(labels = c("t00", "t02", "t04", "t08", "t24"), 
                      values=c("#000000", "#E69F00", "#009E73", "#0072B2", "#D55E00")) +
    geom_text(aes(y = val), label = "*", position = position_dodge(width = 0.7), size = 3) +
    geom_text(aes(y = val01), label = "**", position = position_dodge(width = 0.7), size = 3) +
    theme(title = element_text(size = 16), axis.text = element_text(size = 16)) +
    # theme(legend.title = element_text(size = 16), legend.text = element_text(size = 16)) +
    # theme(legend.position = "bottom")
    theme(legend.position = "none")
  print(p)
  nm2 <- gsub(" ", "_", nm)
  assign(paste("p", nm2, sep="_"),p)
  # ggsave(filename = paste("plots/arcog_",nm2,"_histogram.pdf", sep = ""))
  # ggsave(filename = paste("plots/arcog_",nm2,"_histogram.png", sep = ""))
}

## make legend separately from the plots
# leg <- get_legend(p)
# as_ggplot(leg)


combined_plot <- plot_grid(p_t00_top_10, p_t00_bottom_10, 
                           p_t02_top_10, p_t02_bottom_10, 
                           p_t04_top_10, p_t04_bottom_10, 
                           p_t08_top_10, p_t08_bottom_10, 
                           p_t24_top_10, p_t24_bottom_10, 
          labels = c('A', 'F', 'B', 'G', 'C', 'H', 'D', 'I', 'E', 'J'), label_size = 20, ncol = 2, nrow = 5)
plot_grid(combined_plot, leg, ncol = 1, rel_heights = c(1, .02))
# ggsave("plots/figureS6_top_bottom_10.pdf", height = 36, width = 24, units = "in")
ggsave("plots/figureS7_top_bottom_10_revised.pdf", height = 36, width = 24, units = "in")
ggsave("plots/figureS7_top_bottom_10_revised.eps", height = 36, width = 24, units = "in")
```


```{r clustering, fig.width=15, fig.height=8}
## clustering method suggested by Pariksheet
# https://hbctraining.github.io/DGE_workshop/lessons/08_DGE_LRT.html

load("noiseq/myfactors_mate.RData")

# May need to round count_data again... 
count_data_raw <- read.table("data/mating.mat", header=T, com='', check.names=F)
count_data <- count_data_raw[rowSums(count_data_raw) > 10,]
count_data <- round(count_data, 0)
dds <- DESeqDataSetFromMatrix(countData = count_data, colData = myfactors, design = ~ time_de)

# Wald test
dds_wald <- DESeq(dds, test="Wald")

# Likelihood ratio test
dds_lrt <- DESeq(dds, test="LRT", reduced = ~ 1)
# Extract results
res_LRT <- results(dds_lrt)

# Subset the LRT results to return genes with padj < 0.05
sig_res_LRT <- res_LRT %>%
               data.frame() %>%
               rownames_to_column(var="gene") %>% 
               as_tibble() %>% 
               filter(padj < 0.05)
 
# Get sig gene lists
sigLRT_genes <- sig_res_LRT %>% 
                pull(gene)
                
length(sigLRT_genes)

# Subset results for faster cluster finding (for classroom demo purposes)
clustering_sig_genes <- sig_res_LRT %>%
                  arrange(padj) %>%
                  head(n=1000)

# Transform counts for data visualization
rld <- rlog(dds, blind=TRUE)

# Plot PCA 
plotPCA(rld, intgroup="time")

# Extract the rlog matrix from the object
rld_mat <- assay(rld)

# Compute pairwise correlation values
rld_cor <- cor(rld_mat)

# Plot heatmap
pheatmap(rld_cor)

# Obtain rlog values for those significant genes
cluster_rlog <- rld_mat[clustering_sig_genes$gene, ]

# Use the `degPatterns` function from the 'DEGreport' package to show gene clusters across sample groups
clusters <- degPatterns(cluster_rlog, metadata = myfactors, time = "time_de", col=NULL, minc = 1)

## cluster all genes instead of the top 1000
cluster_rlog_all <- rld_mat[sig_res_LRT$gene, ]
cluster_all <- degPatterns(cluster_rlog_all, metadata = myfactors, time = "time_de", col=NULL, minc = 1)

#### ggplot manually
ggplot(cluster_all$plot$data) +
  geom_line(aes(x = time_de, y = value, group = genes), color = "grey") +
  geom_boxplot(aes(x = time_de, y = value
                   # alpha = 0.2 # transparency isn't working for eps
                   )) +
  geom_point(position = "jitter", aes(x = time_de, y = value 
                                      # alpha = 0.2 # transparency isn't working for eps
                                      )) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(legend.position = "none") +
  ylab("scaled expression") + xlab(NULL) +
  theme(text = element_text(size=24), axis.text.x = element_text(size  = 24, angle = 90, hjust = 1, vjust = 1)) +
  theme(plot.title = element_text(size = 24)) +
  facet_wrap(~cluster, ncol = 4)
ggsave("plots/figS5_clusters_revision.pdf", height = 20, width = 14)
ggsave("plots/figS5_clusters_revision.eps", height = 20, width = 14)
# What type of data structure is the `clusters` output?
class(clusters)

# Let's see what is stored in the `df` component
head(clusters$df)

# Extract the Group 3 genes
cluster_groups <- cluster_all$df
group3 <- cluster_all$df %>%
          filter(cluster == 3)

group12 <- cluster_all$df %>%
          filter(cluster == 12)
```

```{r de on clusters}
#library(annotables)

group3 <- data.frame(lapply(group3, function(x) {
  gsub("lcl.", "lcl|", x)
}))
sub_counts <- subset(annot_counts, ncbi_id %in% group3$genes)

# group12 <- data.frame(lapply(group12, function(x) {
#   gsub("lcl.", "lcl|", x)
# }))
# sub_counts <- subset(annot_counts, ncbi_id %in% group12$genes)

hvo_counts <- sub_counts[, c(4,45:56)]
rownames(hvo_counts) <- NULL
hvo_counts <- distinct(hvo_counts)
rownames(hvo_counts) <- hvo_counts$locus_tag
  hvo_counts$locus_tag <- NULL
  hvo_counts <- data.matrix(hvo_counts)

print(sub_counts$old_locus_tag)
print(rownames(hvo_counts))
list_hvo <- rownames(hvo_counts)

supp_table <- sub_counts[, c(4,2,13,15,37,38)]
supp_table <- distinct(supp_table)
write.table(supp_table, file = "tables/supp_table_cluster3.txt", quote = FALSE, sep = "\t", row.names = FALSE)

## make gene lengths object with HVOs
hvo_gene_lengths <- data.frame("locus_tag" = hvo_map$locus_tag,
                               "length" = hvo_map$length) 
hvo_gene_lengths <- unique(hvo_gene_lengths)
sub_length <- subset(hvo_gene_lengths, locus_tag %in% rownames(hvo_counts))

dds_group12 = NOISeq::readData(data = hvo_counts, factors = myfactors, length = sub_length)
dir.create(path = "noiseq/group3")
time <- c("p0","00","02","04","08","24")
time_de = c("shaking", "t00", "t02", "t04", "t08", "t24")

# Run differential expression for data with biological replicates and construct volcano plots
for(s1 in time_de){
  for(s2 in time_de){
    if(s1!=s2){
      if(s1<s2){
        print(c(s1,s2))
        tmm <- noiseqbio(dds_group12, norm = "tmm", k = 0.5, factor = "time_de", conditions = c(s1,s2), 
                             r = 50, filter = 1, nclust = 15, a0per= 0.9, random.seed = 12345, plot = TRUE, lc = 1)
        save(tmm, file = paste0("noiseq/group3/mate_",s1,"vs",s2,"_tmm.RData"))
        tmm_r <- as.data.frame(tmm@results, row.names = NULL)
        write.table(tmm@results, paste0("noiseq/group3/mate_",s1,"vs",s2,"_tmm.txt"), 
                    sep = "\t", row.names = TRUE, col.names = NA, quote = FALSE)
        tmm_r <- na.omit(tmm_r)
        tmm_r$P <- with(tmm_r, 1-prob)
        tmm_r$locus_tag <- row.names(tmm_r)
        write.table(tmm_r, paste0("noiseq/group3/mate_",s1,"vs",s2,"_tmm_filtered.txt"), 
                    sep = "\t", row.names = TRUE, col.names = NA, quote = FALSE)
        tmm_rm <- merge(tmm_r, hvo_map, by = "locus_tag", all.x=TRUE)
        tmm_rm$EFFECTSIZE <- tmm_rm$log2FC
        write.table(tmm_rm, paste0("noiseq/group3/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"),
                    sep = "\t", row.names = F, quote = FALSE)
      }}}}
## rename cluster number to run on additional clusters (e.g. change "group3" to "group12")
```


```{r group3 volcano}
for(s1 in time_de){
  for(s2 in time_de){
    if(s1!=s2){
      if(s1<s2){
        print(c(s1,s2))
        tmm_r <- read.delim(file = paste0("noiseq/group3/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"), 
                            header=TRUE, row.names = NULL, com='', check.names=F)
        plot(tmm_r$log2FC, -log10(1-tmm_r$prob), pch=20, xlim=c(-7,7), ylim=c(0,16), main = paste(s1,"vs",s2), 
          xlab = "log2FC", ylab = "-log10(p-value)",
            col = ifelse(abs(tmm_r$log2FC)>5 & tmm_r$prob>0.999,"red",
              ifelse(abs(tmm_r$log2FC)>5,"orange",
                ifelse(tmm_r$prob>0.99,"green","black"))))
      }}}}

for(s1 in time_de){
  for(s2 in time_de){
    if(s1!=s2){
      if(s1<s2){
        print(paste0("Expression at time point ",s1," is higher than time point ",s2,
                     " right of zero (i.e. positive) and lower than ",s2, " to the left of zero (i.e. negative)."))
        tmm_r <- read.delim(paste0("noiseq/group3/mate_",s1,"vs",s2,"_tmm_filt_annot.txt"), 
                            header=TRUE, row.names = NULL, com='', check.names=F)
        volc_obj <- volcanor(tmm_r, p = "P", effect_size = "EFFECTSIZE", snp = "locus_tag", 
                             gene = "protein", annotation1 = "locus_tag", annotation2 = "arCOG_name")
        volc_plot <- volcanoly(volc_obj, effect_size_line = c(-1,1), effect_size_line_color = "orange", 
                  genomewideline = -log10(1e-2), genomewideline_color = "green", title = paste(s1,"vs",s2), 
                  highlight = tmm_r$locus_tag, highlight_color = "black")
        print(volc_plot)
        assign(paste("plot", s1, "vs", s2, sep = "_"), volc_plot)
        # ggsave(paste0("plots/volcano_",s1,"vs",s2,"_cluster3.pdf"))
        # ggsave(paste0("plots/volcano_",s1,"vs",s2,"_cluster3.eps"))
      }}}}

plot_shaking_vs_t00
plot_shaking_vs_t02
plot_shaking_vs_t04
plot_shaking_vs_t08
plot_shaking_vs_t24
cls3 <- subplot(plot_shaking_vs_t00, plot_shaking_vs_t02, plot_shaking_vs_t04,
          plot_shaking_vs_t08, plot_shaking_vs_t24, nrows = 3, shareX = TRUE, shareY = TRUE)
# ggsave("plots/figure6_selfish_genes.pdf", height = 15, width = 8, units = "in")
# orca(cls3, "plots/figureS6_cluster3_volcano_revision.pdf", height = 9, width = 6.5)
# need orca command line installed, but currently installed in the wrong place; fix later
```


```{r full supp cluster table}
group <- cluster_all$df %>%
          filter(cluster == 3)

group <- data.frame(lapply(group, function(x) {
  gsub("lcl.", "lcl|", x)
}))
sub_counts <- subset(annot_counts, ncbi_id %in% group$genes)

hvo_counts <- sub_counts[, c(4,45:56)]
rownames(hvo_counts) <- NULL
hvo_counts <- distinct(hvo_counts)
rownames(hvo_counts) <- hvo_counts$locus_tag
  hvo_counts$locus_tag <- NULL
  hvo_counts <- data.matrix(hvo_counts)

print(sub_counts$old_locus_tag)
print(rownames(hvo_counts))
list_hvo <- rownames(hvo_counts)

supp_table <- sub_counts[, c(4,2,13,15,37,38)]
supp_table <- distinct(supp_table)
write.table(supp_table, file = "tables/tableS3_cluster3.txt", quote = FALSE, sep = "\t", row.names = FALSE)

## full excel spreadsheet with additional clusters prepared by combining the above table for each cluster
```

```{r cluster co-located}
## find operons in the cluster data
# filter hvo_map by cluster list -> find start and stops within 100 nt?
# find consecutive by 5 RS within clustered genes
cdf <- NULL
cdf$genes <- cluster_all$raw$genes
cdf$cluster <- cluster_all$raw$cluster
cdf <- as.data.frame(cdf)
cdf <- unique(cdf)
cdf <- data.frame(lapply(cdf, function(x) {
  gsub("lcl.", "lcl|", x)
}))

sub_map <- NULL
sub_map$ncbi_id <- hvo_map$ncbi_id
sub_map$locus_tag <- hvo_map$locus_tag
sub_map$old_locus_tag <- hvo_map$old_locus_tag
sub_map$start <- hvo_map$start
sub_map$stop <- hvo_map$stop
sub_map$chr <- hvo_map$chr
sub_map <- as.data.frame(sub_map)

cdf <- merge(cdf, sub_map, by.x = "genes", by.y = "ncbi_id", all.x = TRUE)

cdf <- separate(cdf, locus_tag, into = c("na", "rs"), sep = 6, remove = FALSE)
cdf <- unique(cdf)
cdf$na <- NULL
cdf$rs <- as.numeric(cdf$rs)
mat <- NULL
# filter by cluster, then for each value in cdf$rs, take abs value of the difference and put into a matrix
dir.create("tables/cluster_operons")
for (x in 1:22) {
  cdfsub <- cdf %>%
          filter(cluster == x)
  cdfsub <- cdfsub[order(cdfsub$rs),]
  cdfsub$op1 <- c(0, abs(diff(cdfsub$rs)) == 5)
  print(cdfsub)
  write.table(cdfsub, file = paste0("tables/cluster_operons/cluster_",x,"_operons.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
  # mat <- matrix(nrow = length(cdfsub$genes), ncol = length(cdfsub$genes))
  # rownames(mat) <- cdfsub$genes
  # colnames(mat) <- cdfsub$genes
  # mat
}

# for i in cluster_*.txt; do awk '/\t1$/{if (a && a !~ /\t1$/) print a; print} {a=$0}' $i > filtered_$i; done
# grep "Chromosome" filtered_cluster_3_operons.txt | awk -F"\t" '{print $3}' > genes_c3.txt
# grep -f genes_c3.txt ../../full_annot.txt | grep -v -i "ribosomal protein" |\
#   awk -F"\t" '{OFS="\t"} {print $1,$4,$7,$8,$13,$15,$38,$42,$43,$45,$46,$47,$48,$49,$50,$51,$52,$53,$54,$55,$56}' |\
#   sort -k 2 | uniq > c3_op_annot.txt
# 

```

```{r gene map}
genes <- read.delim("data/genes_for_map.txt", header = TRUE)
genes_gly <- read.delim("data/genes_for_map_glyc.txt", header = TRUE)

ggplot(genes, aes(xmin = start, xmax = stop, y = chr, fill = locus_tag, label = gene)) +
  geom_gene_arrow() +
  geom_gene_label(align = "left") +
  facet_wrap(~ chr, scales = "free", ncol = 1) +
  scale_fill_brewer(palette = "Set3") +
  theme_genes()

genes$direction <- ifelse(genes$strand == "+", 1, -1)
ggplot(
  subset(genes, chr == "Chromosome"),
  aes(xmin = start, xmax = stop, y = strand, fill = locus_tag, label = gene, forward = direction)
  ) +
  geom_gene_arrow() +
  geom_gene_label() +
  scale_fill_brewer(palette = "Set3") +
  theme_genes() +
  theme(legend.title = element_blank()) +
  theme(legend.key.size = unit(0.3, "cm"))
ggsave("plots/gene_map.pdf", height = 2, width = 6, units = "in")
ggsave("plots/gene_map.eps", height = 2, width = 6, units = "in")

genes_gly$direction <- ifelse(genes_gly$strand == "+", 1, -1)
ggplot(
  subset(genes_gly, chr == "Chromosome"),
  aes(xmin = start, xmax = stop, y = strand, fill = locus_tag, label = gene, forward = direction)
  ) +
  geom_gene_arrow() +
  geom_gene_label() +
  scale_fill_brewer(palette = "Set3") +
  theme_genes() +
  theme(legend.title = element_blank()) +
  theme(legend.key.size = unit(0.3, "cm"))
ggsave("plots/gene_map_glyc.pdf", height = 2, width = 6, units = "in")
```

```{r wet plots, dev='pdf', fig.width=6, fig.height=3}
### from above script
#explo.plot(myPCA, factor = "time")
library(scales)
library(ggplot2)
###
mating <- read.table("data/mating.txt", header=T, com='', check.names=F)
ggplot(mating, aes(time2,average, group=1)) +
  geom_errorbar(aes(ymin=average-stderr,ymax=average+stderr), width=0.1) +
  geom_line() +
  geom_point() +
  scale_x_continuous(breaks = c(0,2,4,8,24), labels = c("0","2","4","8","24")) +
  # scale_x_continuous(breaks = c(0,2,4,8,24), labels = NULL) +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(colour = 'black', size = 0.75),
        text = element_text(size=18, color = "black")
        # axis.ticks.x = element_blank(),
        # axis.ticks.length = unit(3, "mm"),
        # axis.text.x = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.text.y = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.title = element_blank(),
        # axis.text = element_blank()
        ) +
  xlab("Time (h)") +
  ylab("Mean % mated")
ggsave("plots/mean_mated.pdf")

mating <- read.table("data/mating_all.txt", header=T, com='', check.names=F)
ggplot(mating, aes(time2,mated, group=1)) +
  geom_point() +
  scale_x_continuous(breaks = c(0,2,4,8,24), labels = c("0","2","4","8","24")) +
  # scale_x_continuous(breaks = c(0,2,4,8,24), labels = NULL) +
  geom_smooth(method='lm', se = FALSE) +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(colour = 'black', size = 0.75),
        text = element_text(size=18),
        # axis.ticks.x = element_blank(),
        axis.ticks.length = unit(3, "mm"),
        # axis.text.x = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.text.y = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.title = element_blank(),
        axis.text = element_text(color = "black")
        ) +
  scale_y_continuous(labels = scientific) +
  xlab("Time (h)") +
  ylab("Mating efficiency")
ggsave("plots/percent_mated.pdf")

growth <- read.table("data/growth.txt", header=T, com='', check.names=F)
ggplot(growth, aes(time2,cfu_mean, group=1)) +
  geom_errorbar(aes(ymin=cfu_mean-cfu_err,ymax=cfu_mean+cfu_err), width=0.1) +
  geom_point() +
  scale_x_continuous(breaks = c(0,2,4,8,24), labels = c("0","2","4","8","24")) +
  # scale_x_continuous(breaks = c(0,2,4,8,24), labels = NULL) +
  geom_line() +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(colour = 'black', size = 0.75),
        text = element_text(size=18),
        # axis.ticks.x = element_blank(),
        axis.ticks.length = unit(3, "mm"),
        # axis.text.x = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.text.y = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.title = element_blank(),
        axis.text = element_text(color = "black")
        ) +
  scale_y_log10() +
  xlab("time (h)") +
  ylab(substitute(paste("mean CFU", " (log"[10], ")")))
ggsave("plots/growth.pdf")

mating <- read.table("data/mated_full.txt", header=T, com='', check.names=F)
mating_lm <- lm(mated ~ 0 + time, data = mating)
ggplot(mating, aes(time,mated, group=1)) +
  geom_point() +
  scale_x_continuous(breaks = c(-1,0,2,4,8,24), labels = c("sh","0","2","4","8","24")) +
  # scale_x_continuous(breaks = c(0,2,4,8,24), labels = NULL) +
  #geom_smooth(method='glm', se = FALSE) +
  geom_abline(intercept=0, slope=mating_lm$coefficients[1], color='darkred', size=1.1) + 
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(colour = 'black', size = 0.75),
        text = element_text(size=18),
        # axis.ticks.x = element_blank(),
        axis.ticks.length = unit(3, "mm"),
        # axis.text.x = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.text.y = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.title = element_blank(),
        axis.text = element_text(color = "black")
        ) +
  scale_y_continuous(labels = scientific) +
 # scale_y_log10() + 
  xlab("Time (h)") +
  ylab("Mating efficiency")
ggsave("plots/mating_efficiency.pdf")

data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      se = sd(x[[col]], na.rm=TRUE) / sqrt(length(x[[col]])))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
 return(data_sum)
}

mating2 <- data_summary(mating, varname="mated", 
                    groupnames="time")

ggplot(mating2, aes(time, mated, group=1)) +
  geom_point() +
  scale_x_continuous(breaks = c(-1,0,2,4,8,24), labels = c("sh","0","2","4","8","24")) +
  # scale_x_continuous(breaks = c(0,2,4,8,24), labels = NULL) +
  geom_errorbar(aes(ymin=mated-se,ymax=mated+se), width=0.1) +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(colour = 'black', size = 0.75),
        text = element_text(size=18),
        # axis.ticks.x = element_blank(),
        axis.ticks.length = unit(3, "mm"),
        # axis.text.x = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.text.y = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.title = element_blank(),
        axis.text = element_text(color = "black")
        ) +
  scale_y_continuous(labels = scientific) +
  xlab("Time (h)") +
  ylab("Mating efficiency")
ggsave("plots/mating_efficiency_err.pdf")

library(ISwR)
mating_lm <- lm(mated ~ 0 + time, data = mating)
summary(mating_lm)
confint(mating_lm, 'time', level=0.95)

mating_noz <- read.table("data/mated_full_noz.txt", header=T, com='', check.names=F)
mating_lm_noz <- lm(mated ~ time, data = mating_noz)
summary(mating_lm_noz)
confint(mating_lm_noz, 'time', level=0.95)

ggplot(mating_noz, aes(time, mated, group=1)) +
  geom_point(size = 3) +
  scale_x_continuous(breaks = c(0,2,4,8,24), labels = c("","2","4","8","24"), limits = c(0,24)) +
  # scale_x_continuous(breaks = c(0,2,4,8,24), labels = NULL) +
  geom_smooth(method='glm', se = FALSE, size = 2) +
  #geom_abline(intercept=0, slope=mating_lm$coefficients[1], color='darkred', size=1.1) + 
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(colour = 'black', size = 1.25),
        text = element_text(size=18),
        axis.ticks.length = unit(6, "mm"),
        axis.ticks = element_line(size = 1),
        # axis.text.x = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.text.y = element_text(margin=margin(2,2,5,5,"mm")),
        # axis.title = element_blank(),
        axis.text = element_text(color = "black")
        # axis.text=element_blank()
        ) +
  scale_y_continuous(labels = scientific) +
 # scale_y_log10() + 
  xlab("Time (h)") +
  ylab("Mating efficiency")
ggsave("plots/mating_efficiency_nozs.pdf")
```
---
#### Revisions 
Some revisions were made within the above script for minor edits to figures. The following is additional analyses recommended by the reviewers.
```{bash hhpred}
# in silico functional analysis of hypotheticals
# split the protein file for Hvo DS2 to pull out hypothetical proteins
# include qc
mkdir split
mkdir hypothetical
mkdir out
split_multifasta.pl -i ../GCF_000025685.1_ASM2568v1_protein.faa -o split/
grep -c "^>" ../GCF_000025685.1_ASM2568v1_protein.faa
ls split/*.fsa | wc -l
grep -lir "hypothetical" split/* | xargs mv -t hypothetical/
ls split/*.fsa | wc -l
ls hypothetical/*.fsa | wc -l

# 1327 genes have "hypothetical" in the name

# https://github.com/soedinglab/hh-suite/wiki

# installed hhsuite https://github.com/soedinglab/hh-suite using mac os method - no issues
# for reference:
brew install gcc
brew install g++
cd ~/scripts
git clone https://github.com/soedinglab/hh-suite.git
mkdir -p hh-suite/build && cd hh-suite/build
## check version of gcc installed - gcc-10 on 2020.08.25
CC="$(brew --prefix)/bin/gcc-10" CXX="$(brew --prefix)/bin/g++-10" cmake -DCMAKE_INSTALL_PREFIX=. ..
make -j 4 && make install
export PATH="$(pwd)/bin:$(pwd)/scripts:$PATH"

## checking every hypothetical
for i in hypothetical/*.fsa; do 
  file=`basename "${i}"`; echo ${file};
  hhblits -i hypothetical/${file} \
    -o out/${file}.out \
    -oa3m out/${file}.msa \
    -opsi out/${file}.psiblast \
    -d ~/databases/uniclust30_2016_03/uniclust30_2016_03;
done

## focusing on the top up and down
mkdir topbottom
mkdir topbottom_out

for i in /Volumes/LaCie/hvo_mating/tables/de*10.txt; do
  file=`basename "${i}"`; echo ${file};
  awk '{print $1}' /Volumes/LaCie/hvo_mating/tables/${file} |\
  sed 's/.*WP/WP/g; s/1_.*/1/g; 1d' |\
  grep -v "NC_" > ${file}; 
done

for i in *.txt; do
  grep -Fli -f ${i} hypothetical/* | xargs mv -t topbottom/;
done

for i in topbottom/*.fsa; do 
  file=`basename "${i}"`; echo ${file};
  hhblits -i topbottom/${file} \
    -o topbottom_out/${file}.out \
    -oa3m topbottom_out/${file}.msa \
    -opsi topbottom_out/${file}.psiblast \
    -d ~/databases/uniclust30_2016_03/uniclust30_2016_03;
done
```

```{r}

```




## end ##